{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Effect of paralog choice on rRNA full operons\n",
    "- Pseudomonas (also can try on Staph)\n",
    "\n",
    "-> iqtree -s o.fas -wbtl -bb  --> o.fas.uboot will have ultrafast bootsrap trees w blens"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "from mpl_toolkits.axes_grid1 import make_axes_locatable\n",
    "\n",
    "from Bio import Seq, SeqIO, Align, AlignIO, Phylo, Alphabet, pairwise2 \n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Align import AlignInfo, Applications\n",
    "from Bio.Phylo import draw, TreeConstruction  #TreeConstruction.DistanceCalculator, TreeConstruction.DistanceTreeConstructor \n",
    "# https://bioinformatics.stackexchange.com/questions/4337/biopython-phylogenetic-tree-replace-branch-tip-labels-by-sequence-logos\n",
    "\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "from sklearn import manifold, metrics, cluster, neighbors, decomposition, preprocessing\n",
    "import skbio, parasail, dendropy\n",
    "import sys, gzip, re, glob, pickle, collections, subprocess, os, errno, random, itertools\n",
    "\n",
    "def print_redblack(textr, textb=\"\"):\n",
    "    print ('\\x1b[0;1;31;1m'+ str(textr) + '\\x1b[0;1;30;1m'+ str(textb) + '\\x1b[0m')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### defining main variables "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "##  \"Mycobacterium\" is not truely positive or negative\n",
    "gram_positive = [\"Clostridium\", \"Enterococcus\", \"Listeria\", \"Mycobacterium\", \"Staphylococcus\", \"Streptococcus\"]\n",
    "gram_negative = [\"Campylobacter\", \"Escherichia\", \"Helicobacter\", \"Leptospira\", \n",
    "                 \"Neisseria\", \"Pseudomonas\", \"Salmonella\"]\n",
    "#outdir = \"/usr/users/QIB_fr005/deolivl/Academic/Quadram/005.metagenomics_70S/notebooks/20190215.013_results/\"\n",
    "bacteria_genus = \"Staphylococcus\"  # \"Pseudomonas\"\n",
    "outdir = \"./014_results/\" + bacteria_genus\n",
    "rrna_types = [\"16S\", \"23S\", \"5S\"];"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### read refseq genomes "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0;1;31;1mtable_filenames_species[] has:\u001b[0;1;30;1m filename,  GCF_id, full species_subsp name, species name (genus+sp), genus\n",
      "\u001b[0m\n",
      "['/media/deolivl/QIB_deolivl/bigdata/Staphylococcus/GCF_000009005.1_ASM900v1_genomic.gbff.gz', 'GCF_000009005.1', 'Staphylococcus aureus RF122', 'Staphylococcus_aureus', 'Staphylococcus'] \n",
      " {'Staphylococcus'}\n",
      "\u001b[0;1;31;1mdistinct species: \u001b[0;1;30;1mCounter({'Staphylococcus_aureus': 362, 'Staphylococcus_epidermidis': 16, 'Staphylococcus_lugdunensis': 10, 'Staphylococcus_simulans': 7, 'Staphylococcus_saprophyticus': 7, 'Staphylococcus_haemolyticus': 6, 'Staphylococcus_pseudintermedius': 6, 'Staphylococcus_warneri': 5, 'Staphylococcus_schleiferi': 5, 'Staphylococcus_xylosus': 4, 'Staphylococcus_argenteus': 4, 'Staphylococcus_nepalensis': 3, 'Staphylococcus_capitis': 3, 'Staphylococcus_equorum': 3, 'Staphylococcus_condimenti': 3, 'Staphylococcus_hyicus': 3, 'Staphylococcus_caprae': 3, 'Staphylococcus_muscae': 2, 'Staphylococcus_cohnii': 2, 'Staphylococcus_hominis': 2, 'Staphylococcus_carnosus': 2, 'Staphylococcus_sciuri': 2, 'Staphylococcus_pasteuri': 2, 'Staphylococcus_lutrae': 1, 'Staphylococcus_pettenkoferi': 1, 'Staphylococcus_agnetis': 1, 'Staphylococcus_schweitzeri': 1, 'Staphylococcus_stepanovicii': 1, 'Staphylococcus_delphini': 1, 'Staphylococcus_succinus': 1, 'Staphylococcus_auricularis': 1, 'Staphylococcus_kloosii': 1, 'Staphylococcus_simiae': 1, 'Staphylococcus_felis': 1, 'Staphylococcus_piscifermentans': 1})\u001b[0m\n",
      "\u001b[0;1;31;1m{'Staphylococcus_simulans', 'Staphylococcus_capitis', 'Staphylococcus_epidermidis', 'Staphylococcus_warneri', 'Staphylococcus_argenteus', 'Staphylococcus_condimenti', 'Staphylococcus_haemolyticus', 'Staphylococcus_aureus', 'Staphylococcus_hyicus', 'Staphylococcus_saprophyticus', 'Staphylococcus_nepalensis', 'Staphylococcus_lugdunensis', 'Staphylococcus_caprae', 'Staphylococcus_equorum', 'Staphylococcus_schleiferi', 'Staphylococcus_pseudintermedius', 'Staphylococcus_xylosus'}\u001b[0;1;30;1m\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "## read species names translation tables and keep only genus+species \n",
    "file_name = glob.glob(\"/media/deolivl/QIB_deolivl/bigdata/\" + bacteria_genus + \"*/names.txt\")[0]\n",
    "file_lines = [line.strip() for line in open(file_name, 'r')]\n",
    "thistbl = [line.split('\\t') for line in file_lines[1:]] # concatenation of lists\n",
    "fnames = glob.glob(\"/media/deolivl/QIB_deolivl/bigdata/\" + bacteria_genus + \"*/GCF_*.gbff.gz\")\n",
    "idx=[i for j in thistbl for i,fname in enumerate(fnames) if j[0] in fname]\n",
    "table_filenames_species = [[fnames[idx[i]]] + j for i,j in enumerate(thistbl)]\n",
    "table_filenames_species = [line + [\"_\".join(line[2].split()[:2]).split('.')[0]] for line in table_filenames_species]\n",
    "table_filenames_species = [line + [line[2].split()[0]] for line in table_filenames_species]\n",
    "print_redblack (\"table_filenames_species[] has:\",\" filename,  GCF_id, full species_subsp name, species name (genus+sp), genus\\n\")\n",
    "\n",
    "print (table_filenames_species[0], \"\\n\", set([x[4] for x  in table_filenames_species]))\n",
    "\n",
    "species_counter = collections.Counter([x[3] for x in table_filenames_species if x[3][-3:] !=\"_sp\"])\n",
    "print_redblack (\"distinct species: \", species_counter)\n",
    "\n",
    "def return_species_name_from_table (table_fs):\n",
    "    return str(table_fs[3]) + \"_\" + str(re.search('GCF_(\\d+)\\.', table_fs[1]).group(1))\n",
    "\n",
    "tbl_fl_sp = [x  for x in table_filenames_species if species_counter[ x[3] ] > 2]\n",
    "print_redblack(set([x[3] for x  in tbl_fl_sp]))\n",
    "random.shuffle(tbl_fl_sp)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### main functions\n",
    "* find all operons, align sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def align_seqs (sequences=None, maxiters=12, infile=None, outfile=None, mafft = True):\n",
    "    print (\"started aligning...\", flush=True, end=\" \")\n",
    "    if (sequences is None) and (infile is None):\n",
    "        print (\"ERROR: You must give me an alignment object or file\")\n",
    "        return [] ## OTOH if both are present then infile is overwritten with contents of sequences[]\n",
    "    if infile is None:\n",
    "        ifl = \"/tmp/in.fas\"\n",
    "    else:\n",
    "        ifl = infile\n",
    "    if outfile is None:\n",
    "        ofl = \"/tmp/out.fas\"\n",
    "    else:\n",
    "        ofl = outfile\n",
    "    SeqIO.write(sequences, ifl, \"fasta\")\n",
    "    if (mafft is False):\n",
    "        proc_run = subprocess.check_output(\"muscle -in \" + ifl + \" -diags -maxiters \" + str(maxiters) + \" -out \" + ofl,\n",
    "                                       shell=True, universal_newlines=True)\n",
    "    else: # \"--parttree --6merpair\" and 0.123 is to avoid very long alignments\n",
    "        proc_run = subprocess.check_output(\"mafft --ep 0.3 --op 3.0 --auto \" + ifl + \" > \" + ofl,\n",
    "                                       shell=True, universal_newlines=True)\n",
    "      \n",
    "    aligned = AlignIO.read(ofl, \"fasta\")\n",
    "    print (\"Finished\",flush=True)\n",
    "    if infile is None:\n",
    "        os.system(\"rm -f \" + ifl)\n",
    "    if outfile is None:\n",
    "        os.system(\"rm -f \" + ofl)\n",
    "    return aligned\n",
    "\n",
    "def return_contig_locations (featurelist): # featurelist has list of [ [location1, location2], rnaname, strand ]\n",
    "    featurelist = sorted(featurelist, key = lambda x: int(x[0][0]))\n",
    "    rnablocks = []\n",
    "    rnabnames = []\n",
    "    rnasignal = []\n",
    "    tmp_loc = list(featurelist[0][0]) # start and end locations\n",
    "    tmp_nam = [featurelist[0][1]]    # gene name\n",
    "    tmp_sig = [featurelist[0][2]]\n",
    "    for feature in featurelist[1:]:\n",
    "        if (abs(tmp_loc[-1] - feature[0][0]) < 1000):\n",
    "            tmp_loc.extend(list(feature[0]))\n",
    "            tmp_nam.extend([feature[1]])\n",
    "            tmp_sig.extend([feature[2]])\n",
    "        else:\n",
    "            rnablocks.append(tmp_loc)\n",
    "            rnabnames.append(tmp_nam)\n",
    "            rnasignal.append(tmp_sig)\n",
    "            tmp_loc = list(feature[0])\n",
    "            tmp_nam = [feature[1]]\n",
    "            tmp_sig = [feature[2]]\n",
    "    rnablocks.append(tmp_loc)\n",
    "    rnabnames.append(tmp_nam)\n",
    "    rnasignal.append(tmp_sig)\n",
    "    return rnablocks, rnabnames, rnasignal\n",
    "    \n",
    "def get_rrna_from_genbank_to_list_and_dict (table_fs):\n",
    "    gbank = SeqIO.parse(gzip.open(table_fs[0], \"rt\"), \"genbank\")\n",
    "    list_of_features = []\n",
    "    list_of_seqs = []\n",
    "    genome = next(gbank) # no point in iterating over gbank, only this has whole information\n",
    "    for feature in genome.features:\n",
    "        if(feature.type == \"rRNA\"):\n",
    "            for rna in rrna_types:\n",
    "                if rna in feature.qualifiers['product'][0]:\n",
    "                    this_product = rna\n",
    "                    break\n",
    "            this_location = [int(i) for i in re.findall('\\d+',str(feature.location))]\n",
    "            this_strand = feature.strand\n",
    "            if 0 not in this_location:\n",
    "                list_of_features.append([this_location[:2], this_product, this_strand])\n",
    "    # now list_of_features is complete  for this genome record (gbank[0])\n",
    "    if not list_of_features:\n",
    "        print (\"no info from \", table_fs[0])\n",
    "        return None # some genomes have no relevant annotation\n",
    "    location, gnames, strand = return_contig_locations (list_of_features)\n",
    "    genome_len = len(genome.seq)\n",
    "    for i, (l,g,s) in enumerate(zip(location, gnames, strand)):\n",
    "        i_start = min(l); i_end = max(l)\n",
    "        if sum(s) < 0: ## hopefully they all have same signal\n",
    "            contig = genome.seq[i_start:i_end].reverse_complement()\n",
    "            seqname = \"s\" + \"\".join(g[::-1])\n",
    "        else:\n",
    "            contig = genome.seq[i_start:i_end]\n",
    "            seqname = \"s\" + \"\".join(g)\n",
    "        numbercode = '_{:02d}'.format(i)\n",
    "        seqname = return_species_name_from_table(table_fs) + numbercode\n",
    "        if (len(contig) > 1000) and (len(contig) < 8500): # some contigs have only 5S; some genomes have only this 5S annotated\n",
    "            list_of_seqs.append(SeqRecord(contig, id = seqname, description = seqname))\n",
    "    return list_of_seqs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### sample genomes and create unaligned list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20 40 60 80 100 # first sequence of operon alignment:\n",
      " ID: Staphylococcus_aureus_900324415_01\n",
      "Name: <unknown name>\n",
      "Description: Staphylococcus_aureus_900324415_01\n",
      "Number of features: 0\n",
      "Seq('TTTATGGAGAGTTTGATCCTGGCTCAGGATGAACGCTGGCGGCGTGCCTAATAC...GGC', IUPACAmbiguousDNA())\n",
      "# percentiles: 4924.0 6082.52 442\n"
     ]
    }
   ],
   "source": [
    "good_fl_sp = []\n",
    "rna_operon = []  # includes all paralogs\n",
    "for file_counter, tfs in enumerate(tbl_fl_sp[:100]):\n",
    "    if not (file_counter+1)%20:\n",
    "        print (str(file_counter+1), end=\" \", flush=True)\n",
    "    rna_list = get_rrna_from_genbank_to_list_and_dict (tfs)\n",
    "    # (some list will have no usable features or b/c too long/too short; some dict has no usable features\n",
    "    if rna_list:\n",
    "        good_fl_sp.append(tfs)\n",
    "        rna_operon.extend(rna_list)\n",
    "\n",
    "# save table, remembering that 'outdir' has the genus name already\n",
    "fl=gzip.open(outdir+\"_fl_sp.pickle.gz\", \"w\"); pickle.dump([good_fl_sp],fl,2); fl.close()\n",
    "\n",
    "print (\"# first sequence of operon alignment:\\n\", rna_operon[0])\n",
    "x = [len(i.seq) for i in rna_operon]\n",
    "print (\"# percentiles:\", np.percentile(x, 5), np.percentile(x, 99), len(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "started aligning... Finished\n"
     ]
    }
   ],
   "source": [
    "# align each of the dictionary values (lists of unaligned seqs)\n",
    "operon_aligned = align_seqs(sequences=rna_operon, outfile=outdir + \".fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_species_genus_labels (seqlist, have_paralogs = False): \n",
    "    species = ['_'.join(sequence.id.split('_')[:2]) for sequence in seqlist]\n",
    "    genus = [sequence.id.split('_')[0] for sequence in seqlist]\n",
    "    if have_paralogs:# assumes 'genus_species_code_number' names \n",
    "        sample = ['_'.join(sequence.id.split('_')[:3]) for sequence in seqlist]\n",
    "        return species, genus, sample\n",
    "    return species, genus "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### if resuming from previous analysis, run code below to read genome names table and alignments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read table\n",
    "fl=gzip.open(outdir+\"_fl_sp.pickle.gz\", \"r\"); good_fl_sp = pickle.load(fl); fl.close()\n",
    "operon_aligned = AlignIO.read(outdir + \".fasta\", \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "IQ-TREE multicore version 1.3.11.1 for Linux 64-bit built Feb  1 2016\n",
      "Copyright (c) 2011-2015 Nguyen Lam Tung, Olga Chernomor, Arndt von Haeseler and Bui Quang Minh.\n",
      "\n",
      "Host:    n114312.nbi.ac.uk (AVX2, FMA3, 15 GB RAM)\n",
      "Command: iqtree-omp -s ./014_results/Staphylococcus.fasta -nt 8 -m HKY+G\n",
      "Seed:    882015 (Using SPRNG - Scalable Parallel Random Number Generator)\n",
      "Time:    Fri Feb 22 15:48:14 2019\n",
      "Kernel:  AVX - 8 threads (8 CPU cores detected)\n",
      "\n",
      "Reading alignment file ./014_results/Staphylococcus.fasta ... Fasta format detected\n",
      "Alignment most likely contains DNA/RNA sequences\n",
      "WARNING: 1 sites contain only gaps or ambiguous characters.\n",
      "Alignment has 442 sequences with 7296 columns and 1456 patterns (986 informative sites)\n",
      "                                          Gap/Ambiguity  Composition  p-value\n",
      "   1  Staphylococcus_aureus_900324415_01         31.29%    passed     99.85%\n",
      "   2  Staphylococcus_aureus_900324415_02         31.30%    passed     99.87%\n",
      "   3  Staphylococcus_aureus_900324415_03         29.40%    passed     90.34%\n",
      "   4  Staphylococcus_aureus_900324415_04         28.50%    passed     99.92%\n",
      "   5  Staphylococcus_aureus_900324285_00         31.29%    passed     99.81%\n",
      "   6  Staphylococcus_aureus_900324285_01         29.40%    passed     91.49%\n",
      "   7  Staphylococcus_aureus_900324285_02         28.52%    passed     99.71%\n",
      "   8  Staphylococcus_aureus_900324285_04         31.29%    passed     99.85%\n",
      "   9  Staphylococcus_aureus_003354645_00         31.09%    passed     98.66%\n",
      "  10  Staphylococcus_aureus_003354645_02         29.59%    passed     92.66%\n",
      "  11  Staphylococcus_aureus_003354645_03         29.63%    passed     93.86%\n",
      "  12  Staphylococcus_aureus_003354645_04         31.09%    passed     99.34%\n",
      "  13  Staphylococcus_aureus_000967405_00         29.61%    passed     96.33%\n",
      "  14  Staphylococcus_aureus_000967405_01         16.93%    passed     42.29%\n",
      "  15  Staphylococcus_aureus_000967405_02         29.59%    passed     95.64%\n",
      "  16  Staphylococcus_aureus_000967405_03         29.63%    passed     91.89%\n",
      "  17  Staphylococcus_aureus_000967405_04         29.63%    passed     92.62%\n",
      "  18  Staphylococcus_argenteus_002812345_00      30.85%    passed     97.65%\n",
      "  19  Staphylococcus_argenteus_002812345_01      16.43%    passed     45.15%\n",
      "  20  Staphylococcus_argenteus_002812345_02      29.55%    passed     97.53%\n",
      "  21  Staphylococcus_argenteus_002812345_03      31.95%    passed     99.89%\n",
      "  22  Staphylococcus_argenteus_002812345_04      30.59%    passed     98.13%\n",
      "  23  Staphylococcus_aureus_900323965_00         30.21%    passed     81.60%\n",
      "  24  Staphylococcus_aureus_900323965_01         30.26%    passed     99.53%\n",
      "  25  Staphylococcus_aureus_900323965_02         31.07%    passed     99.58%\n",
      "  26  Staphylococcus_aureus_900323965_03         31.32%    passed     99.93%\n",
      "  27  Staphylococcus_aureus_900635265_00         31.09%    passed     99.06%\n",
      "  28  Staphylococcus_aureus_900635265_02         29.58%    passed     97.16%\n",
      "  29  Staphylococcus_aureus_900635265_03         30.25%    passed     96.25%\n",
      "  30  Staphylococcus_aureus_900635265_04         28.59%    passed     74.90%\n",
      "  31  Staphylococcus_xylosus_002078255_00        29.52%    passed     98.47%\n",
      "  32  Staphylococcus_xylosus_002078255_01        32.03%    passed     98.20%\n",
      "  33  Staphylococcus_xylosus_002078255_02        32.03%    passed     98.20%\n",
      "  34  Staphylococcus_epidermidis_002085695_00    32.54%    passed     70.02%\n",
      "  35  Staphylococcus_epidermidis_002085695_01    32.51%    passed     75.85%\n",
      "  36  Staphylococcus_epidermidis_002085695_02    29.93%    passed     55.59%\n",
      "  37  Staphylococcus_epidermidis_002085695_04    32.52%    passed     73.69%\n",
      "  38  Staphylococcus_aureus_000024585_00         31.07%    passed     99.21%\n",
      "  39  Staphylococcus_aureus_000024585_01         18.12%    passed     58.24%\n",
      "  40  Staphylococcus_aureus_000024585_02         29.58%    passed     95.95%\n",
      "  41  Staphylococcus_aureus_000024585_03         29.63%    passed     91.09%\n",
      "  42  Staphylococcus_aureus_000024585_04         29.63%    passed     92.62%\n",
      "  43  Staphylococcus_aureus_002000785_00         31.07%    passed     98.94%\n",
      "  44  Staphylococcus_aureus_002000785_01         18.11%    passed     55.48%\n",
      "  45  Staphylococcus_aureus_002000785_02         29.58%    passed     97.42%\n",
      "  46  Staphylococcus_aureus_002000785_03         29.58%    passed     96.51%\n",
      "  47  Staphylococcus_aureus_002000785_04         31.92%    passed     98.46%\n",
      "  48  Staphylococcus_aureus_002355355_00         31.07%    passed     99.10%\n",
      "  49  Staphylococcus_aureus_002355355_02         29.58%    passed     97.78%\n",
      "  50  Staphylococcus_aureus_002355355_03         30.22%    passed     96.17%\n",
      "  51  Staphylococcus_aureus_002355355_04         28.58%    passed     75.66%\n",
      "  52  Staphylococcus_haemolyticus_000009865_00   31.48%    passed     98.22%\n",
      "  53  Staphylococcus_haemolyticus_000009865_01   30.10%    passed     89.30%\n",
      "  54  Staphylococcus_haemolyticus_000009865_02   29.92%    passed     83.70%\n",
      "  55  Staphylococcus_haemolyticus_000009865_03   17.96%    passed     53.97%\n",
      "  56  Staphylococcus_haemolyticus_000009865_04   31.50%    passed     98.81%\n",
      "  57  Staphylococcus_aureus_001515705_00         31.09%    passed     99.57%\n",
      "  58  Staphylococcus_aureus_001515705_01         18.08%    passed     54.18%\n",
      "  59  Staphylococcus_aureus_001515705_02         29.52%    passed     98.39%\n",
      "  60  Staphylococcus_aureus_001515705_03         28.56%    passed     76.95%\n",
      "  61  Staphylococcus_aureus_001515705_04         28.59%    passed     73.93%\n",
      "  62  Staphylococcus_aureus_000239235_00         31.92%    passed     98.18%\n",
      "  63  Staphylococcus_aureus_000239235_01         18.38%    passed     59.74%\n",
      "  64  Staphylococcus_aureus_000239235_02         29.58%    passed     95.78%\n",
      "  65  Staphylococcus_aureus_000239235_03         30.03%    passed     92.67%\n",
      "  66  Staphylococcus_aureus_000239235_04         30.26%    passed     98.82%\n",
      "  67  Staphylococcus_aureus_003288395_00         31.09%    passed     99.06%\n",
      "  68  Staphylococcus_aureus_003288395_01         30.26%    passed     99.44%\n",
      "  69  Staphylococcus_aureus_003288395_02         31.09%    passed     99.25%\n",
      "  70  Staphylococcus_aureus_003288395_03         29.59%    passed     94.74%\n",
      "  71  Staphylococcus_aureus_003354865_00         29.65%    passed     94.45%\n",
      "  72  Staphylococcus_aureus_003354865_02         28.54%    passed     99.99%\n",
      "  73  Staphylococcus_aureus_003354865_03         28.59%    passed     73.93%\n",
      "  74  Staphylococcus_aureus_003354865_04         28.59%    passed     73.93%\n",
      "  75  Staphylococcus_aureus_003193705_00         31.11%    passed     98.72%\n",
      "  76  Staphylococcus_aureus_003193705_01         18.16%    passed     59.13%\n",
      "  77  Staphylococcus_aureus_003193705_02         29.61%    passed     96.23%\n",
      "  78  Staphylococcus_aureus_003193705_03         29.74%    passed     90.86%\n",
      "  79  Staphylococcus_aureus_003193705_04         29.67%    passed     92.09%\n",
      "  80  Staphylococcus_aureus_900478245_00         30.28%    passed     99.32%\n",
      "  81  Staphylococcus_aureus_900478245_01         16.45%    passed     48.54%\n",
      "  82  Staphylococcus_aureus_900478245_02         29.58%    passed     94.58%\n",
      "  83  Staphylococcus_aureus_900478245_03         31.06%    passed     99.75%\n",
      "  84  Staphylococcus_aureus_900478245_04         30.04%    passed     99.39%\n",
      "  85  Staphylococcus_warneri_900636265_00        30.93%    passed     97.49%\n",
      "  86  Staphylococcus_warneri_900636265_01        30.95%    passed     97.90%\n",
      "  87  Staphylococcus_warneri_900636265_02        30.28%    passed     86.94%\n",
      "  88  Staphylococcus_warneri_900636265_04        30.83%    passed     95.16%\n",
      "  89  Staphylococcus_aureus_000967345_00         29.58%    passed     96.72%\n",
      "  90  Staphylococcus_aureus_000967345_01         18.12%    passed     59.43%\n",
      "  91  Staphylococcus_aureus_000967345_02         29.58%    passed     94.78%\n",
      "  92  Staphylococcus_aureus_000967345_03         29.63%    passed     91.89%\n",
      "  93  Staphylococcus_aureus_000967345_04         29.63%    passed     92.62%\n",
      "  94  Staphylococcus_warneri_003691405_00        30.96%    passed     97.26%\n",
      "  95  Staphylococcus_warneri_003691405_01        30.96%    passed     97.25%\n",
      "  96  Staphylococcus_warneri_003691405_02        30.29%    passed     85.22%\n",
      "  97  Staphylococcus_warneri_003691405_03        18.28%    passed     65.41%\n",
      "  98  Staphylococcus_warneri_003691405_04        30.98%    passed     97.31%\n",
      "  99  Staphylococcus_aureus_002386245_00         31.11%    passed     98.87%\n",
      " 100  Staphylococcus_aureus_002386245_02         29.59%    passed     97.73%\n",
      " 101  Staphylococcus_aureus_002386245_03         28.59%    passed     75.44%\n",
      " 102  Staphylococcus_aureus_002386245_04         28.60%    passed     72.84%\n",
      " 103  Staphylococcus_lugdunensis_000025085_00    29.47%    passed     96.58%\n",
      " 104  Staphylococcus_lugdunensis_000025085_01    28.50%    passed     90.52%\n",
      " 105  Staphylococcus_lugdunensis_000025085_02    28.11%    passed     96.90%\n",
      " 106  Staphylococcus_lugdunensis_000025085_04    28.50%    passed     88.82%\n",
      " 107  Staphylococcus_caprae_003966625_00         32.51%    passed     57.84%\n",
      " 108  Staphylococcus_caprae_003966625_01         32.52%    passed     58.47%\n",
      " 109  Staphylococcus_caprae_003966625_02         30.48%    passed     30.78%\n",
      " 110  Staphylococcus_caprae_003966625_03         18.54%    passed     27.03%\n",
      " 111  Staphylococcus_caprae_003966625_04         32.52%    passed     55.10%\n",
      " 112  Staphylococcus_aureus_002089095_00         29.65%    passed     95.78%\n",
      " 113  Staphylococcus_aureus_002089095_02         28.36%    passed     72.27%\n",
      " 114  Staphylococcus_aureus_002089095_03         29.41%    passed     92.07%\n",
      " 115  Staphylococcus_aureus_002089095_04         31.29%    passed     99.87%\n",
      " 116  Staphylococcus_aureus_003193725_00         31.09%    passed     98.92%\n",
      " 117  Staphylococcus_aureus_003193725_02         29.59%    passed     97.30%\n",
      " 118  Staphylococcus_aureus_003193725_03         31.94%    passed     98.09%\n",
      " 119  Staphylococcus_aureus_003193725_04         31.94%    passed     98.22%\n",
      " 120  Staphylococcus_aureus_003347055_00         31.72%    passed     98.25%\n",
      " 121  Staphylococcus_aureus_003347055_02         28.54%    passed     99.99%\n",
      " 122  Staphylococcus_aureus_003347055_03         28.37%    passed     68.19%\n",
      " 123  Staphylococcus_aureus_003347055_04         28.59%    passed     72.98%\n",
      " 124  Staphylococcus_aureus_001641025_00         31.07%    passed     99.10%\n",
      " 125  Staphylococcus_aureus_001641025_01         18.09%    passed     56.66%\n",
      " 126  Staphylococcus_aureus_001641025_02         29.58%    passed     97.19%\n",
      " 127  Staphylococcus_aureus_001641025_03         28.58%    passed     77.48%\n",
      " 128  Staphylococcus_aureus_001641025_04         28.58%    passed     69.81%\n",
      " 129  Staphylococcus_aureus_000953255_00         29.58%    passed     94.92%\n",
      " 130  Staphylococcus_aureus_000953255_01         29.58%    passed     94.94%\n",
      " 131  Staphylococcus_aureus_000953255_02         17.72%    passed     50.07%\n",
      " 132  Staphylococcus_aureus_000953255_03         31.06%    passed     99.68%\n",
      " 133  Staphylococcus_aureus_000953255_04         30.03%    passed     99.04%\n",
      " 134  Staphylococcus_aureus_002089055_00         29.63%    passed     95.50%\n",
      " 135  Staphylococcus_aureus_002089055_02         28.51%    passed     99.72%\n",
      " 136  Staphylococcus_aureus_002089055_03         29.39%    passed     91.26%\n",
      " 137  Staphylococcus_aureus_002089055_04         31.28%    passed     99.85%\n",
      " 138  Staphylococcus_aureus_002795285_00         30.24%    passed     98.49%\n",
      " 139  Staphylococcus_aureus_002795285_01         16.86%    passed     51.85%\n",
      " 140  Staphylococcus_aureus_002795285_02         28.50%    passed    100.00%\n",
      " 141  Staphylococcus_aureus_002795285_03         28.51%    passed     73.48%\n",
      " 142  Staphylococcus_aureus_002795285_04         31.94%    passed     98.48%\n",
      " 143  Staphylococcus_condimenti_900637235_00     32.74%    passed     17.81%\n",
      " 144  Staphylococcus_condimenti_900637235_01     35.21%    passed      7.43%\n",
      " 145  Staphylococcus_condimenti_900637235_02     30.19%    passed     16.08%\n",
      " 146  Staphylococcus_condimenti_900637235_04     32.76%    passed     14.94%\n",
      " 147  Staphylococcus_aureus_001021895_00         31.07%    passed     98.94%\n",
      " 148  Staphylococcus_aureus_001021895_02         29.58%    passed     97.42%\n",
      " 149  Staphylococcus_aureus_001021895_03         31.92%    passed     98.31%\n",
      " 150  Staphylococcus_aureus_001021895_04         31.92%    passed     98.46%\n",
      " 151  Staphylococcus_warneri_003571725_00        30.95%    passed     96.11%\n",
      " 152  Staphylococcus_warneri_003571725_01        30.93%    passed     97.13%\n",
      " 153  Staphylococcus_warneri_003571725_02        30.30%    passed     84.32%\n",
      " 154  Staphylococcus_warneri_003571725_04        30.93%    passed     97.32%\n",
      " 155  Staphylococcus_aureus_003186105_00         30.28%    passed     98.58%\n",
      " 156  Staphylococcus_aureus_003186105_02         30.03%    passed     91.76%\n",
      " 157  Staphylococcus_aureus_003186105_03         29.59%    passed     97.32%\n",
      " 158  Staphylococcus_aureus_003186105_04         31.94%    passed     99.08%\n",
      " 159  Staphylococcus_aureus_900474695_00         31.09%    passed     99.41%\n",
      " 160  Staphylococcus_aureus_900474695_02         28.52%    passed     99.94%\n",
      " 161  Staphylococcus_aureus_900474695_03         30.25%    passed     95.27%\n",
      " 162  Staphylococcus_aureus_900474695_04         31.09%    passed     99.34%\n",
      " 163  Staphylococcus_haemolyticus_002952715_00   31.52%    passed     98.33%\n",
      " 164  Staphylococcus_haemolyticus_002952715_01   31.52%    passed     98.69%\n",
      " 165  Staphylococcus_haemolyticus_002952715_03   29.93%    passed     81.02%\n",
      " 166  Staphylococcus_haemolyticus_002952715_04   30.14%    passed     92.47%\n",
      " 167  Staphylococcus_aureus_002633865_00         31.07%    passed     98.98%\n",
      " 168  Staphylococcus_aureus_002633865_02         29.58%    passed     96.94%\n",
      " 169  Staphylococcus_aureus_002633865_03         29.63%    passed     92.62%\n",
      " 170  Staphylococcus_aureus_002633865_04         29.63%    passed     93.28%\n",
      " 171  Staphylococcus_aureus_003354905_00         31.09%    passed     98.66%\n",
      " 172  Staphylococcus_aureus_003354905_02         29.58%    passed     93.18%\n",
      " 173  Staphylococcus_aureus_003354905_03         29.63%    passed     93.86%\n",
      " 174  Staphylococcus_aureus_003354905_04         31.09%    passed     99.01%\n",
      " 175  Staphylococcus_aureus_001611365_00         31.10%    passed     99.23%\n",
      " 176  Staphylococcus_aureus_001611365_02         31.09%    passed     99.08%\n",
      " 177  Staphylococcus_aureus_001611365_03         31.07%    passed     98.93%\n",
      " 178  Staphylococcus_aureus_001611365_04         29.65%    passed     92.17%\n",
      " 179  Staphylococcus_lugdunensis_002591215_01    29.45%    passed     96.58%\n",
      " 180  Staphylococcus_lugdunensis_002591215_02    29.45%    passed     95.89%\n",
      " 181  Staphylococcus_lugdunensis_002591215_03    29.48%    passed     93.22%\n",
      " 182  Staphylococcus_lugdunensis_002591215_04    27.11%    passed     99.27%\n",
      " 183  Staphylococcus_aureus_002633845_00         31.07%    passed     99.36%\n",
      " 184  Staphylococcus_aureus_002633845_02         29.58%    passed     94.92%\n",
      " 185  Staphylococcus_aureus_002633845_03         31.07%    passed     99.27%\n",
      " 186  Staphylococcus_aureus_002633845_04         30.24%    passed     99.38%\n",
      " 187  Staphylococcus_aureus_002442975_00         31.07%    passed     99.26%\n",
      " 188  Staphylococcus_aureus_002442975_02         29.58%    passed     95.80%\n",
      " 189  Staphylococcus_aureus_002442975_03         31.70%    passed     98.54%\n",
      " 190  Staphylococcus_aureus_002442975_04         30.26%    passed     99.44%\n",
      " 191  Staphylococcus_aureus_002000605_00         31.07%    passed     98.94%\n",
      " 192  Staphylococcus_aureus_002000605_01         18.11%    passed     55.48%\n",
      " 193  Staphylococcus_aureus_002000605_02         29.58%    passed     97.42%\n",
      " 194  Staphylococcus_aureus_002000605_03         29.58%    passed     96.51%\n",
      " 195  Staphylococcus_aureus_002000605_04         31.92%    passed     98.46%\n",
      " 196  Staphylococcus_aureus_000452385_01         23.40%    passed     10.01%\n",
      " 197  Staphylococcus_aureus_000452385_02         28.58%    passed     75.96%\n",
      " 198  Staphylococcus_aureus_000452385_03         78.66%    passed     27.38%\n",
      " 199  Staphylococcus_aureus_000452385_04         57.44%    passed     73.03%\n",
      " 200  Staphylococcus_aureus_000452385_05         28.59%    passed     76.42%\n",
      " 201  Staphylococcus_aureus_003193785_00         28.59%    passed     68.83%\n",
      " 202  Staphylococcus_aureus_003193785_01         28.37%    passed     67.52%\n",
      " 203  Staphylococcus_aureus_003193785_02         28.54%    passed     99.94%\n",
      " 204  Staphylococcus_aureus_003193785_04         31.72%    passed     98.11%\n",
      " 205  Staphylococcus_aureus_900324295_00         28.52%    passed     99.57%\n",
      " 206  Staphylococcus_aureus_900324295_01         29.40%    passed     91.49%\n",
      " 207  Staphylococcus_aureus_900324295_02         31.29%    passed     99.86%\n",
      " 208  Staphylococcus_aureus_900324295_03         31.29%    passed     99.85%\n",
      " 209  Staphylococcus_aureus_002000865_00         31.07%    passed     98.94%\n",
      " 210  Staphylococcus_aureus_002000865_01         18.11%    passed     55.48%\n",
      " 211  Staphylococcus_aureus_002000865_02         29.58%    passed     97.42%\n",
      " 212  Staphylococcus_aureus_002000865_03         29.58%    passed     96.51%\n",
      " 213  Staphylococcus_aureus_002000865_04         31.92%    passed     98.46%\n",
      " 214  Staphylococcus_aureus_001717975_00         29.59%    passed     95.83%\n",
      " 215  Staphylococcus_aureus_001717975_02         31.09%    passed     98.02%\n",
      " 216  Staphylococcus_aureus_001717975_03         31.94%    passed     98.67%\n",
      " 217  Staphylococcus_aureus_001717975_04         31.94%    passed     98.09%\n",
      " 218  Staphylococcus_aureus_003595505_00         31.09%    passed     98.92%\n",
      " 219  Staphylococcus_aureus_003595505_01         18.11%    passed     55.48%\n",
      " 220  Staphylococcus_aureus_003595505_02         29.59%    passed     97.30%\n",
      " 221  Staphylococcus_aureus_003595505_03         29.59%    passed     96.38%\n",
      " 222  Staphylococcus_aureus_003595505_04         31.94%    passed     98.22%\n",
      " 223  Staphylococcus_aureus_002140115_00         31.70%    passed     98.65%\n",
      " 224  Staphylococcus_aureus_002140115_02         29.58%    passed     96.09%\n",
      " 225  Staphylococcus_aureus_002140115_03         28.58%    passed     72.78%\n",
      " 226  Staphylococcus_aureus_002140115_04         28.58%    passed     71.49%\n",
      " 227  Staphylococcus_aureus_001018775_00         29.59%    passed     96.10%\n",
      " 228  Staphylococcus_aureus_001018775_01         29.65%    passed     92.24%\n",
      " 229  Staphylococcus_aureus_001018775_02         29.65%    passed     92.96%\n",
      " 230  Staphylococcus_aureus_001018775_03         31.09%    passed     98.82%\n",
      " 231  Staphylococcus_aureus_001021875_00         31.07%    passed     99.21%\n",
      " 232  Staphylococcus_aureus_001021875_02         29.58%    passed     97.42%\n",
      " 233  Staphylococcus_aureus_001021875_03         31.92%    passed     98.62%\n",
      " 234  Staphylococcus_aureus_001021875_04         31.92%    passed     98.46%\n",
      " 235  Staphylococcus_aureus_000144955_00         31.07%    passed     99.19%\n",
      " 236  Staphylococcus_aureus_000144955_01         17.72%    passed     54.65%\n",
      " 237  Staphylococcus_aureus_000144955_02         31.07%    passed     99.42%\n",
      " 238  Staphylococcus_aureus_000144955_03         31.06%    passed     99.39%\n",
      " 239  Staphylococcus_aureus_000144955_04         31.07%    passed     99.40%\n",
      " 240  Staphylococcus_lugdunensis_002250095_00    29.45%    passed     94.73%\n",
      " 241  Staphylococcus_lugdunensis_002250095_01    29.45%    passed     93.54%\n",
      " 242  Staphylococcus_lugdunensis_002250095_02    27.11%    passed     99.27%\n",
      " 243  Staphylococcus_lugdunensis_002250095_03    16.23%    passed     79.13%\n",
      " 244  Staphylococcus_lugdunensis_002250095_04    16.47%    passed     76.65%\n",
      " 245  Staphylococcus_lugdunensis_002250095_05    29.45%    passed     96.58%\n",
      " 246  Staphylococcus_epidermidis_900086615_00    32.51%    passed     77.56%\n",
      " 247  Staphylococcus_epidermidis_900086615_01    18.20%    passed     34.50%\n",
      " 248  Staphylococcus_epidermidis_900086615_02    29.95%    passed     59.59%\n",
      " 249  Staphylococcus_epidermidis_900086615_03    32.52%    passed     78.19%\n",
      " 250  Staphylococcus_epidermidis_900086615_04    32.54%    passed     71.83%\n",
      " 251  Staphylococcus_aureus_003073715_00         31.09%    passed     98.89%\n",
      " 252  Staphylococcus_aureus_003073715_02         29.59%    passed     96.10%\n",
      " 253  Staphylococcus_aureus_003073715_03         29.65%    passed     92.24%\n",
      " 254  Staphylococcus_aureus_003073715_04         29.65%    passed     92.96%\n",
      " 255  Staphylococcus_aureus_900636695_00         30.25%    passed     94.08%\n",
      " 256  Staphylococcus_aureus_900636695_01         15.21%    passed     31.35%\n",
      " 257  Staphylococcus_aureus_900636695_02         29.58%    passed     95.58%\n",
      " 258  Staphylococcus_aureus_900636695_03         30.03%    passed     87.04%\n",
      " 259  Staphylococcus_aureus_900636695_04         30.25%    passed     95.51%\n",
      " 260  Staphylococcus_epidermidis_900636255_00    30.95%    passed     96.28%\n",
      " 261  Staphylococcus_epidermidis_900636255_01    30.95%    passed     95.67%\n",
      " 262  Staphylococcus_epidermidis_900636255_02    30.28%    passed     84.55%\n",
      " 263  Staphylococcus_epidermidis_900636255_04    30.95%    passed     96.45%\n",
      " 264  Staphylococcus_aureus_003944865_00         30.28%    passed     99.40%\n",
      " 265  Staphylococcus_aureus_003944865_01         18.78%    passed     51.23%\n",
      " 266  Staphylococcus_aureus_003944865_02         29.59%    passed     94.74%\n",
      " 267  Staphylococcus_aureus_003944865_03         31.09%    passed     99.52%\n",
      " 268  Staphylococcus_aureus_003944865_04         30.06%    passed     99.23%\n",
      " 269  Staphylococcus_aureus_003184985_00         30.03%    passed     90.91%\n",
      " 270  Staphylococcus_aureus_003184985_02         29.59%    passed     94.80%\n",
      " 271  Staphylococcus_aureus_003184985_03         28.36%    passed     70.33%\n",
      " 272  Staphylococcus_aureus_003184985_04         31.07%    passed     99.24%\n",
      " 273  Staphylococcus_aureus_003111745_00         29.65%    passed     94.18%\n",
      " 274  Staphylococcus_aureus_003111745_02         28.52%    passed     99.71%\n",
      " 275  Staphylococcus_aureus_003111745_03         29.40%    passed     92.39%\n",
      " 276  Staphylococcus_aureus_003111745_04         31.29%    passed     99.93%\n",
      " 277  Staphylococcus_epidermidis_000759555_00    29.96%    passed     62.19%\n",
      " 278  Staphylococcus_epidermidis_000759555_01    32.55%    passed     78.25%\n",
      " 279  Staphylococcus_epidermidis_000759555_02    32.55%    passed     74.75%\n",
      " 280  Staphylococcus_epidermidis_000759555_03    32.50%    passed     75.91%\n",
      " 281  Staphylococcus_aureus_002000725_00         31.07%    passed     98.94%\n",
      " 282  Staphylococcus_aureus_002000725_01         18.11%    passed     55.48%\n",
      " 283  Staphylococcus_aureus_002000725_02         29.58%    passed     97.42%\n",
      " 284  Staphylococcus_aureus_002000725_03         29.58%    passed     96.51%\n",
      " 285  Staphylococcus_aureus_002000725_04         31.92%    passed     98.46%\n",
      " 286  Staphylococcus_aureus_003855575_00         30.28%    passed     99.56%\n",
      " 287  Staphylococcus_aureus_003855575_01         16.93%    passed     50.29%\n",
      " 288  Staphylococcus_aureus_003855575_02         29.65%    passed     94.45%\n",
      " 289  Staphylococcus_aureus_003855575_03         31.07%    passed     99.75%\n",
      " 290  Staphylococcus_aureus_003855575_04         30.04%    passed     99.21%\n",
      " 291  Staphylococcus_epidermidis_003856455_01    32.54%    passed     76.11%\n",
      " 292  Staphylococcus_epidermidis_003856455_02    32.54%    passed     71.80%\n",
      " 293  Staphylococcus_epidermidis_003856455_03    32.58%    passed     75.77%\n",
      " 294  Staphylococcus_epidermidis_003856455_04    29.98%    passed     62.61%\n",
      " 295  Staphylococcus_warneri_000332735_00        30.91%    passed     96.59%\n",
      " 296  Staphylococcus_warneri_000332735_01        30.91%    passed     96.86%\n",
      " 297  Staphylococcus_warneri_000332735_02        30.91%    passed     96.59%\n",
      " 298  Staphylococcus_warneri_000332735_03        17.71%    passed     79.21%\n",
      " 299  Staphylococcus_warneri_000332735_04        30.91%    passed     96.57%\n",
      " 300  Staphylococcus_aureus_001465635_00         29.63%    passed     93.44%\n",
      " 301  Staphylococcus_aureus_001465635_02         28.51%    passed     99.63%\n",
      " 302  Staphylococcus_aureus_001465635_03         29.39%    passed     92.09%\n",
      " 303  Staphylococcus_aureus_001465635_04         31.28%    passed     99.95%\n",
      " 304  Staphylococcus_hyicus_900474585_00         28.51%    passed     74.15%\n",
      " 305  Staphylococcus_hyicus_900474585_01         29.76%    passed     76.22%\n",
      " 306  Staphylococcus_hyicus_900474585_02         26.29%    passed     71.21%\n",
      " 307  Staphylococcus_hyicus_900474585_04         30.24%    passed     68.13%\n",
      " 308  Staphylococcus_aureus_003627835_00         30.25%    passed     94.51%\n",
      " 309  Staphylococcus_aureus_003627835_01         16.91%    passed     50.29%\n",
      " 310  Staphylococcus_aureus_003627835_02         29.33%    passed     94.35%\n",
      " 311  Staphylococcus_aureus_003627835_03         31.02%    passed     98.57%\n",
      " 312  Staphylococcus_aureus_003627835_04         34.59%    passed     97.28%\n",
      " 313  Staphylococcus_aureus_002204575_00         31.29%    passed     99.91%\n",
      " 314  Staphylococcus_aureus_002204575_02         28.51%    passed     99.63%\n",
      " 315  Staphylococcus_aureus_002204575_03         29.43%    passed     90.53%\n",
      " 316  Staphylococcus_aureus_002204575_04         31.29%    passed     99.90%\n",
      " 317  Staphylococcus_aureus_003952625_01         31.95%    passed     98.13%\n",
      " 318  Staphylococcus_aureus_003952625_02         30.29%    passed     98.55%\n",
      " 319  Staphylococcus_aureus_003952625_03         28.63%    passed     73.52%\n",
      " 320  Staphylococcus_aureus_003952625_04         28.52%    passed     99.94%\n",
      " 321  Staphylococcus_simulans_003096155_00       32.77%    passed     29.81%\n",
      " 322  Staphylococcus_simulans_003096155_01       35.25%    passed     21.38%\n",
      " 323  Staphylococcus_simulans_003096155_02       30.37%    passed     30.84%\n",
      " 324  Staphylococcus_simulans_003096155_04       32.76%    passed     34.41%\n",
      " 325  Staphylococcus_aureus_000695215_00         30.02%    passed     92.70%\n",
      " 326  Staphylococcus_aureus_000695215_02         29.58%    passed     93.40%\n",
      " 327  Staphylococcus_aureus_000695215_03         28.34%    passed     67.27%\n",
      " 328  Staphylococcus_aureus_000695215_04         31.06%    passed     99.22%\n",
      " 329  Staphylococcus_aureus_000626615_00         28.59%    passed     73.31%\n",
      " 330  Staphylococcus_aureus_000626615_01         28.59%    passed     76.42%\n",
      " 331  Staphylococcus_aureus_000626615_02         29.59%    passed     97.85%\n",
      " 332  Staphylococcus_aureus_000626615_04         31.09%    passed     99.06%\n",
      " 333  Staphylococcus_aureus_002633825_00         31.92%    passed     98.37%\n",
      " 334  Staphylococcus_aureus_002633825_02         28.52%    passed     99.96%\n",
      " 335  Staphylococcus_aureus_002633825_03         30.24%    passed     96.13%\n",
      " 336  Staphylococcus_aureus_002633825_04         30.29%    passed     98.39%\n",
      " 337  Staphylococcus_aureus_003595445_00         31.09%    passed     98.92%\n",
      " 338  Staphylococcus_aureus_003595445_01         18.11%    passed     55.48%\n",
      " 339  Staphylococcus_aureus_003595445_02         29.59%    passed     97.30%\n",
      " 340  Staphylococcus_aureus_003595445_03         29.59%    passed     96.38%\n",
      " 341  Staphylococcus_aureus_003595445_04         31.94%    passed     98.22%\n",
      " 342  Staphylococcus_aureus_001717725_00         31.94%    passed     98.22%\n",
      " 343  Staphylococcus_aureus_001717725_01         31.94%    passed     95.33%\n",
      " 344  Staphylococcus_aureus_001717725_02         29.59%    passed     96.69%\n",
      " 345  Staphylococcus_aureus_001717725_04         31.09%    passed     99.57%\n",
      " 346  Staphylococcus_aureus_003426145_00         28.59%    passed     74.90%\n",
      " 347  Staphylococcus_aureus_003426145_01         28.60%    passed     71.91%\n",
      " 348  Staphylococcus_aureus_003426145_02         29.62%    passed     96.44%\n",
      " 349  Staphylococcus_aureus_003426145_04         31.14%    passed     98.19%\n",
      " 350  Staphylococcus_aureus_000815085_00         31.10%    passed     98.68%\n",
      " 351  Staphylococcus_aureus_000815085_01         27.03%    passed     95.49%\n",
      " 352  Staphylococcus_aureus_000815085_02         29.58%    passed     97.42%\n",
      " 353  Staphylococcus_aureus_000815085_03         28.39%    passed     67.52%\n",
      " 354  Staphylococcus_aureus_000815085_04         28.59%    passed     74.56%\n",
      " 355  Staphylococcus_aureus_000463055_00         69.35%    passed     49.77%\n",
      " 356  Staphylococcus_aureus_000463055_01         17.75%    passed     53.80%\n",
      " 357  Staphylococcus_aureus_000463055_02         31.10%    passed     99.17%\n",
      " 358  Staphylococcus_aureus_000463055_03         28.37%    passed     69.08%\n",
      " 359  Staphylococcus_aureus_000463055_04         28.59%    passed     70.27%\n",
      " 360  Staphylococcus_aureus_000737615_00         30.84%    passed     98.97%\n",
      " 361  Staphylococcus_aureus_000737615_01         17.64%    passed     51.11%\n",
      " 362  Staphylococcus_aureus_000737615_02         29.58%    passed     95.93%\n",
      " 363  Staphylococcus_aureus_000737615_03         31.07%    passed     99.27%\n",
      " 364  Staphylococcus_aureus_000737615_04         31.07%    passed     99.59%\n",
      " 365  Staphylococcus_aureus_003194405_00         31.32%    passed     99.67%\n",
      " 366  Staphylococcus_aureus_003194405_01         18.12%    passed     58.13%\n",
      " 367  Staphylococcus_aureus_003194405_02         29.59%    passed     96.10%\n",
      " 368  Staphylococcus_aureus_003194405_03         29.65%    passed     92.24%\n",
      " 369  Staphylococcus_aureus_003194405_04         29.65%    passed     92.96%\n",
      " 370  Staphylococcus_aureus_900324315_00         28.52%    passed     99.71%\n",
      " 371  Staphylococcus_aureus_900324315_01         29.39%    passed     91.80%\n",
      " 372  Staphylococcus_aureus_900324315_02         31.29%    passed     99.86%\n",
      " 373  Staphylococcus_aureus_900324315_03         31.29%    passed     99.85%\n",
      " 374  Staphylococcus_aureus_003573835_00         31.94%    passed     98.18%\n",
      " 375  Staphylococcus_aureus_003573835_02         28.54%    passed     99.97%\n",
      " 376  Staphylococcus_aureus_003573835_03         28.59%    passed     77.58%\n",
      " 377  Staphylococcus_aureus_003573835_04         30.28%    passed     98.55%\n",
      " 378  Staphylococcus_aureus_003030085_00         29.59%    passed     94.33%\n",
      " 379  Staphylococcus_aureus_003030085_01         31.37%    passed     99.18%\n",
      " 380  Staphylococcus_aureus_003030085_02         30.28%    passed     99.47%\n",
      " 381  Staphylococcus_aureus_003030085_03         28.59%    passed     73.88%\n",
      " 382  Staphylococcus_aureus_001515685_00         30.80%    passed     99.24%\n",
      " 383  Staphylococcus_aureus_001515685_01         18.11%    passed     57.63%\n",
      " 384  Staphylococcus_aureus_001515685_02         29.56%    passed     97.08%\n",
      " 385  Staphylococcus_aureus_001515685_03         28.48%    passed     74.77%\n",
      " 386  Staphylococcus_aureus_001515685_04         28.55%    passed     69.79%\n",
      " 387  Staphylococcus_aureus_002000805_00         31.07%    passed     98.94%\n",
      " 388  Staphylococcus_aureus_002000805_01         18.11%    passed     55.48%\n",
      " 389  Staphylococcus_aureus_002000805_02         29.58%    passed     97.42%\n",
      " 390  Staphylococcus_aureus_002000805_03         29.58%    passed     96.51%\n",
      " 391  Staphylococcus_aureus_002000805_04         31.92%    passed     98.46%\n",
      " 392  Staphylococcus_aureus_000470865_00         30.84%    passed     98.97%\n",
      " 393  Staphylococcus_aureus_000470865_01         17.65%    passed     51.85%\n",
      " 394  Staphylococcus_aureus_000470865_02         29.58%    passed     95.80%\n",
      " 395  Staphylococcus_aureus_000470865_03         31.09%    passed     99.35%\n",
      " 396  Staphylococcus_aureus_000470865_04         30.26%    passed     99.44%\n",
      " 397  Staphylococcus_aureus_001019495_00         29.59%    passed     94.50%\n",
      " 398  Staphylococcus_aureus_001019495_01         18.13%    passed     53.99%\n",
      " 399  Staphylococcus_aureus_001019495_02         31.09%    passed     99.01%\n",
      " 400  Staphylococcus_aureus_001019495_03         29.65%    passed     92.96%\n",
      " 401  Staphylococcus_aureus_001019495_04         29.65%    passed     92.96%\n",
      " 402  Staphylococcus_aureus_003595365_00         31.09%    passed     98.92%\n",
      " 403  Staphylococcus_aureus_003595365_01         18.11%    passed     55.71%\n",
      " 404  Staphylococcus_aureus_003595365_02         29.59%    passed     97.30%\n",
      " 405  Staphylococcus_aureus_003595365_03         29.59%    passed     96.38%\n",
      " 406  Staphylococcus_aureus_003595365_04         31.94%    passed     98.22%\n",
      " 407  Staphylococcus_aureus_000462955_00         31.10%    passed     98.30%\n",
      " 408  Staphylococcus_aureus_000462955_01         18.59%    passed     56.72%\n",
      " 409  Staphylococcus_aureus_000462955_02         30.84%    passed     98.60%\n",
      " 410  Staphylococcus_aureus_000462955_03         30.24%    passed     96.22%\n",
      " 411  Staphylococcus_aureus_000462955_04         31.92%    passed     97.87%\n",
      " 412  Staphylococcus_aureus_900004855_00         30.26%    passed     99.45%\n",
      " 413  Staphylococcus_aureus_900004855_01         16.87%    passed     38.10%\n",
      " 414  Staphylococcus_aureus_900004855_02         30.21%    passed     78.98%\n",
      " 415  Staphylococcus_aureus_900004855_03         30.04%    passed     98.28%\n",
      " 416  Staphylococcus_aureus_900004855_04         30.04%    passed     95.52%\n",
      " 417  Staphylococcus_aureus_003203655_01         31.09%    passed     98.82%\n",
      " 418  Staphylococcus_aureus_003203655_02         29.65%    passed     92.09%\n",
      " 419  Staphylococcus_aureus_003203655_03         29.65%    passed     92.24%\n",
      " 420  Staphylococcus_aureus_003203655_04         29.59%    passed     96.44%\n",
      " 421  Staphylococcus_aureus_001018725_00         28.59%    passed     73.31%\n",
      " 422  Staphylococcus_aureus_001018725_01         31.09%    passed     99.07%\n",
      " 423  Staphylococcus_aureus_001018725_02         31.09%    passed     99.07%\n",
      " 424  Staphylococcus_aureus_001018725_04         29.59%    passed     95.03%\n",
      " 425  Staphylococcus_aureus_002000625_00         31.07%    passed     98.94%\n",
      " 426  Staphylococcus_aureus_002000625_01         18.11%    passed     55.48%\n",
      " 427  Staphylococcus_aureus_002000625_02         29.58%    passed     97.42%\n",
      " 428  Staphylococcus_aureus_002000625_03         29.58%    passed     96.51%\n",
      " 429  Staphylococcus_aureus_002000625_04         31.92%    passed     98.46%\n",
      " 430  Staphylococcus_aureus_900635315_00         30.24%    passed     94.24%\n",
      " 431  Staphylococcus_aureus_900635315_02         29.59%    passed     95.68%\n",
      " 432  Staphylococcus_aureus_900635315_03         30.25%    passed     95.23%\n",
      " 433  Staphylococcus_aureus_900635315_04         30.25%    passed     95.51%\n",
      " 434  Staphylococcus_aureus_000013465_00         31.07%    passed     99.07%\n",
      " 435  Staphylococcus_aureus_000013465_01         27.01%    passed     96.45%\n",
      " 436  Staphylococcus_aureus_000013465_02         29.58%    passed     97.82%\n",
      " 437  Staphylococcus_aureus_000013465_03         28.58%    passed     74.21%\n",
      " 438  Staphylococcus_aureus_000013465_04         28.58%    passed     74.41%\n",
      " 439  Staphylococcus_aureus_001045795_00         31.07%    passed     98.73%\n",
      " 440  Staphylococcus_aureus_001045795_02         29.58%    passed     97.42%\n",
      " 441  Staphylococcus_aureus_001045795_03         31.92%    passed     98.31%\n",
      " 442  Staphylococcus_aureus_001045795_04         31.92%    passed     98.46%\n",
      "WARNING: 3 sequences contain more than 50% gaps/ambiguity\n",
      "****  TOTAL                                      29.38%  0 sequences failed composition chi2 test (p-value<5%; df=3)\n",
      "NOTE: 83 identical sequences will be ignored during tree search\n",
      "\n",
      "Create initial parsimony tree by phylogenetic likelihood library (PLL)... 0.093 seconds\n",
      "\n",
      "NOTE: 65 MB RAM is required!\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "1. Initial log-likelihood: -31912.255\n",
      "2. Current log-likelihood: -29935.380\n",
      "3. Current log-likelihood: -29874.798\n",
      "4. Current log-likelihood: -29871.638\n",
      "Optimal log-likelihood: -29871.551\n",
      "Rate parameters:  A-C: 1.000  A-G: 2.394  A-T: 1.000  C-G: 1.000  C-T: 2.394  G-T: 1.000\n",
      "Base frequencies:   A: 0.282  C: 0.204  G: 0.282  T: 0.233\n",
      "Gamma shape alpha: 0.157\n",
      "Parameters optimization took 4 rounds (1.265 sec)\n",
      "\n",
      "--------------------------------------------------------------------\n",
      "|             INITIALIZING CANDIDATE TREE SET                      |\n",
      "--------------------------------------------------------------------\n",
      "Computing ML distances based on estimated model parameters... 1.856 sec\n",
      "WARNING: Some pairwise ML distances are too long (saturated)\n",
      "Computing BIONJ tree...\n",
      "0.085 seconds\n",
      "Log-likelihood of BIONJ tree: -30476.587\n",
      "Generating 98 parsimony trees... 10.146 seconds 100 distinct starting trees\n",
      "Computing log-likelihood of 100 initial trees ... 11.630 seconds\n",
      "Optimizing top 20 initial trees with NNI...\n",
      "Iteration 1 / LogL: -29747.192 / Time: 0h:0m:24s\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "BETTER TREE FOUND at iteration 1: -29747.099\n",
      "Iteration 2 / LogL: -29742.868 / Time: 0h:0m:25s\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "BETTER TREE FOUND at iteration 2: -29742.846\n",
      "Iteration 3 / LogL: -29735.319 / Time: 0h:0m:26s\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "BETTER TREE FOUND at iteration 3: -29735.225\n",
      "Iteration 4 / LogL: -29752.769 / Time: 0h:0m:27s\n",
      "Iteration 5 / LogL: -29753.903 / Time: 0h:0m:27s\n",
      "Iteration 6 / LogL: -29761.303 / Time: 0h:0m:28s\n",
      "Iteration 7 / LogL: -29762.815 / Time: 0h:0m:29s\n",
      "Iteration 8 / LogL: -29746.760 / Time: 0h:0m:29s\n",
      "Iteration 9 / LogL: -29750.272 / Time: 0h:0m:30s\n",
      "Iteration 10 / LogL: -29744.868 / Time: 0h:0m:31s\n",
      "Iteration 11 / LogL: -29772.486 / Time: 0h:0m:32s\n",
      "Iteration 12 / LogL: -29770.630 / Time: 0h:0m:33s\n",
      "Iteration 13 / LogL: -29758.020 / Time: 0h:0m:34s\n",
      "Iteration 14 / LogL: -29758.822 / Time: 0h:0m:35s\n",
      "Iteration 15 / LogL: -29772.359 / Time: 0h:0m:36s\n",
      "Iteration 16 / LogL: -29764.239 / Time: 0h:0m:37s\n",
      "Iteration 17 / LogL: -29763.126 / Time: 0h:0m:38s\n",
      "Iteration 18 / LogL: -29791.478 / Time: 0h:0m:39s\n",
      "Iteration 19 / LogL: -29780.533 / Time: 0h:0m:40s\n",
      "Iteration 20 / LogL: -29745.686 / Time: 0h:0m:40s\n",
      "Average CPU time for 1 NNI search: 6.460\n",
      "Finish initializing candidate tree set. Number of distinct locally optimal trees: 20\n",
      "Current best tree score: -29735.225 / CPU time: 231.244\n",
      "--------------------------------------------------------------------\n",
      "|               OPTIMIZING CANDIDATE TREE SET                      |\n",
      "--------------------------------------------------------------------\n",
      "Iteration 30 / LogL: -29797.417 / Time: 0h:0m:51s (0h:2m:6s left)\n",
      "Iteration 40 / LogL: -29771.913 / Time: 0h:1m:2s (0h:1m:36s left)\n",
      "Iteration 50 / LogL: -29971.028 / Time: 0h:1m:12s (0h:1m:15s left)\n",
      "Iteration 60 / LogL: -29961.094 / Time: 0h:1m:24s (0h:0m:58s left)\n",
      "Iteration 70 / LogL: -29781.599 / Time: 0h:1m:34s (0h:0m:42s left)\n",
      "Iteration 80 / LogL: -29927.602 / Time: 0h:1m:45s (0h:0m:28s left)\n",
      "Iteration 90 / LogL: -29793.908 / Time: 0h:1m:57s (0h:0m:14s left)\n",
      "Iteration 100 / LogL: -29772.685 / Time: 0h:2m:8s (0h:0m:1s left)\n",
      "TREE SEARCH COMPLETED AFTER 101 ITERATIONS / Time: 0h:2m:9s\n",
      "\n",
      "--------------------------------------------------------------------\n",
      "|                    FINALIZING TREE SEARCH                        |\n",
      "--------------------------------------------------------------------\n",
      "Performs final model parameters optimization\n",
      "Estimate model parameters (epsilon = 0.010)\n",
      "1. Initial log-likelihood: -29735.221\n",
      "2. Current log-likelihood: -29735.197\n",
      "Optimal log-likelihood: -29735.192\n",
      "Rate parameters:  A-C: 1.000  A-G: 2.433  A-T: 1.000  C-G: 1.000  C-T: 2.433  G-T: 1.000\n",
      "Base frequencies:   A: 0.282  C: 0.204  G: 0.282  T: 0.233\n",
      "Gamma shape alpha: 0.157\n",
      "Parameters optimization took 2 rounds (0.460 sec)\n",
      "\n",
      "BEST SCORE FOUND : -29735.192\n",
      "Total tree length: 0.973\n",
      "\n",
      "Total number of iterations: 101\n",
      "CPU time used for tree search: 920.528 sec (0h:15m:20s)\n",
      "Wall-clock time used for tree search: 126.826 sec (0h:2m:6s)\n",
      "Total CPU time used: 939.980 sec (0h:15m:39s)\n",
      "Total wall-clock time used: 129.580 sec (0h:2m:9s)\n",
      "\n",
      "Analysis results written to: \n",
      "  IQ-TREE report:                ./014_results/Staphylococcus.fasta.iqtree\n",
      "  Maximum-likelihood tree:       ./014_results/Staphylococcus.fasta.treefile\n",
      "  Likelihood distances:          ./014_results/Staphylococcus.fasta.mldist\n",
      "  Screen log file:               ./014_results/Staphylococcus.fasta.log\n",
      "\n",
      "Date and Time: Fri Feb 22 15:50:24 2019\n",
      "\n"
     ]
    }
   ],
   "source": [
    "outcheck = subprocess.check_output(\"muscle -maketree -in \" + outdir + \".fasta -out \" + outdir + \"_upgma.treefile\", shell=True,universal_newlines=True)\n",
    "outcheck = subprocess.check_output(\"iqtree-omp -s \" + outdir + \".fasta -nt 8 -m HKY+G\",shell=True,universal_newlines=True)\n",
    "print (outcheck)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def patristic_distances_from_treefile (filename, have_paralogs = False, shuffle = False):\n",
    "    tree = dendropy.Tree.get(path=filename, schema=\"newick\", preserve_underscores=True)\n",
    "    if shuffle:\n",
    "        tree = dendropy.simulate.treesim.pure_kingman_tree(taxon_namespace=tree.taxon_namespace, pop_size = 1000)\n",
    "    species = ['_'.join(t.label.split('_')[:2]) for t in tree.taxon_namespace] ## follow order of taxon_namespace\n",
    "    genus = [t.label.split('_')[0] for t in tree.taxon_namespace]\n",
    "    ntaxa = len(tree.taxon_namespace)\n",
    "    distmat = np.zeros((ntaxa,ntaxa)) # diagonals are zero\n",
    "    nodemat = np.zeros((ntaxa,ntaxa))\n",
    "    pdm = tree.phylogenetic_distance_matrix() # initialises class\n",
    "    for i,j in itertools.combinations(range(ntaxa),2):\n",
    "        distmat[i,j] = distmat[j,i] = pdm.distance(tree.taxon_namespace[i], tree.taxon_namespace[j])\n",
    "        nodemat[i,j] = nodemat[j,i] = pdm.path_edge_count(tree.taxon_namespace[i], tree.taxon_namespace[j])\n",
    "    if have_paralogs:\n",
    "        sample = ['_'.join(t.label.split('_')[:3]) for t in tree.taxon_namespace]\n",
    "        return species, genus, distmat, nodemat, sample\n",
    "    return species, genus, distmat, nodemat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def silhouette_str(dist, labels):\n",
    "    distrib = metrics.silhouette_samples(dist, labels, metric=\"precomputed\")\n",
    "    return '{:7.3f}'.format(np.percentile(distrib, 5)) + \\\n",
    "    '{:7.3f}'.format(np.percentile(distrib, 25)) + \\\n",
    "    '{:7.3f}  '.format(sum(distrib>0)/float(len(distrib))) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Each triplet from `silhouette_str()` has $5\\%$, $25\\%$ and proportion of positive silhouettes. And we have two triplets on the left for _species_ (S), and two on the right for _strain_ (T), one being from weighted (W) and one from unweighted (U) patristic distances. Therefore the header is: SW SU TW TU\n",
    "\n",
    "By the way having low silhouettte scores is not necessarily bad, since\n",
    "1. clusters of size one will have score of _zero_ **by definition** (should be _one_ given the silhouette function but this would lead to ever-increasing clusterings)\n",
    "2. clusters with within-distance zero are perfect, while for phylogenies we want non-zero diversification\n",
    "3. they can be negative if for a case like `((A:a,A:a,A:a),B:b)` with $a>b$ since $d(A,A)=2a > a+b = d(A,B)$ --- however the taxonomic classification is fine for A and B"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "random \t  -0.445 -0.398  0.007    -0.302 -0.223  0.005    |   -0.936 -0.904  0.023    -0.524 -0.438  0.020  \n",
      "upgma  \t  -0.001  0.106  0.948    -0.542 -0.284  0.267    |   -0.756 -0.533  0.090    -0.713 -0.588  0.102  \n",
      "ML     \t  -0.291 -0.206  0.731    -0.558 -0.367  0.183    |   -0.872 -0.782  0.118    -0.615 -0.474  0.143  \n"
     ]
    }
   ],
   "source": [
    "sp, ge, mat, mat2, strain = patristic_distances_from_treefile (outdir + \".fasta.treefile\", have_paralogs = True, shuffle = True)\n",
    "print (\"random \\t\", silhouette_str(mat, sp),  silhouette_str(mat2, sp), \" | \", \n",
    "      silhouette_str(mat, strain),  silhouette_str(mat2, strain))\n",
    "\n",
    "for suffix1, title1 in zip([\"_upgma.treefile\",\".fasta.treefile\"], [\"upgma \", \"ML    \"]):\n",
    "    ## operon doenst have _long or _consensus\n",
    "    sp, ge, mat, mat2, strain = patristic_distances_from_treefile (outdir + suffix1, have_paralogs = True)\n",
    "    print (title1, \"\\t\", silhouette_str(mat, sp),  silhouette_str(mat2, sp), \" | \", \n",
    "      silhouette_str(mat, strain),  silhouette_str(mat2, strain))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "random  \t   2.264  4.000  0.368\t   8.275 17.100  0.000\t\n",
      "upgma   \t   1.218  2.000  0.843\t  14.423 45.000  0.073\t\n",
      "ML      \t   1.718  2.000  0.893\t  12.759 28.000  0.091\t\n"
     ]
    }
   ],
   "source": [
    "def sp_count_from_list (labels):\n",
    "    x = ['_'.join(t.split('_')[:2]) for t in labels]\n",
    "    return len(set(x))\n",
    "\n",
    "def strain_count_from_list (labels):\n",
    "    x = ['_'.join(t.split('_')[:3]) for t in labels]\n",
    "    return len(set(x))\n",
    "\n",
    "def bipartitions_from_treefile (filename, shuffle = False):\n",
    "    tree = dendropy.Tree.get(path=filename, schema=\"newick\", preserve_underscores=True)\n",
    "    if shuffle:\n",
    "        tree = dendropy.simulate.treesim.pure_kingman_tree(taxon_namespace=tree.taxon_namespace, pop_size = 1000)\n",
    "    tree.encode_bipartitions()\n",
    "    # bits[i] have the list of labels associated to bipartition (edge) i\n",
    "    bits = [[x.label for x in tree.taxon_namespace.bitmask_taxa_list(i.leafset_bitmask)] for i in tree.bipartition_encoding]\n",
    "    sp_count = [sp_count_from_list(i) for i in bits if len(i) > 1]\n",
    "    strain_count = [strain_count_from_list(i) for i in bits if len(i) > 1]\n",
    "    return sp_count, strain_count\n",
    "\n",
    "def bipart_str(counts):\n",
    "    distrib = np.array(counts)\n",
    "    return '{:7.3f}'.format(np.mean(distrib)) + \\\n",
    "    '{:7.3f}'.format(np.percentile(distrib, 90)) + \\\n",
    "    '{:7.3f}\\t'.format(sum(distrib == 1)/float(len(distrib))) \n",
    "\n",
    "sp0, ge0 = bipartitions_from_treefile (outdir + \".fasta.treefile\", shuffle = True)\n",
    "print ('{:7s}'.format(\"random\"), \"\\t\", bipart_str(sp0), bipart_str(ge0))\n",
    "\n",
    "for suffix1, title1 in zip([\"_upgma.treefile\",\".fasta.treefile\"], [\"upgma \", \"ML    \"]):\n",
    "    sp, ge = bipartitions_from_treefile (outdir + suffix1)    \n",
    "    print ('{:7s}'.format(title1), \"\\t\", bipart_str(sp), bipart_str(ge))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
