{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Effect of paralog choice on rRNA full operons\n",
    "- Pseudomonas (also can try on Staph)\n",
    "\n",
    "-> iqtree -s o.fas -wbtl -bb  --> o.fas.uboot will have ultrafast bootsrap trees w blens"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "from mpl_toolkits.axes_grid1 import make_axes_locatable\n",
    "\n",
    "from Bio import Seq, SeqIO, Align, AlignIO, Phylo, Alphabet, pairwise2 \n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Align import AlignInfo, Applications\n",
    "from Bio.Phylo import draw, TreeConstruction  #TreeConstruction.DistanceCalculator, TreeConstruction.DistanceTreeConstructor \n",
    "# https://bioinformatics.stackexchange.com/questions/4337/biopython-phylogenetic-tree-replace-branch-tip-labels-by-sequence-logos\n",
    "\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "from sklearn import manifold, metrics, cluster, neighbors, decomposition, preprocessing\n",
    "import skbio, parasail, dendropy\n",
    "import sys, gzip, re, glob, pickle, collections, subprocess, os, errno, random, itertools\n",
    "\n",
    "def print_redblack(textr, textb=\"\"):\n",
    "    print ('\\x1b[0;1;31;1m'+ str(textr) + '\\x1b[0;1;30;1m'+ str(textb) + '\\x1b[0m')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### defining main variables "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "##  \"Mycobacterium\" is not truely positive or negative\n",
    "gram_positive = [\"Clostridium\", \"Enterococcus\", \"Listeria\", \"Mycobacterium\", \"Staphylococcus\", \"Streptococcus\"]\n",
    "gram_negative = [\"Campylobacter\", \"Escherichia\", \"Helicobacter\", \"Leptospira\", \n",
    "                 \"Neisseria\", \"Pseudomonas\", \"Salmonella\"]\n",
    "#outdir = \"/usr/users/QIB_fr005/deolivl/Academic/Quadram/005.metagenomics_70S/notebooks/20190215.013_results/\"\n",
    "bacteria_genus = \"Pseudomonas\"  #  \"Staphylococcus\"\n",
    "outdir = \"./014_results/\" + bacteria_genus\n",
    "rrna_types = [\"16S\", \"23S\", \"5S\"];"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### read refseq genomes "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0;1;31;1mtable_filenames_species[] has:\u001b[0;1;30;1m filename,  GCF_id, full species_subsp name, species name (genus+sp), genus\n",
      "\u001b[0m\n",
      "['/media/deolivl/QIB_deolivl/bigdata/Pseudomonas/GCF_000006765.1_ASM676v1_genomic.gbff.gz', 'GCF_000006765.1', 'Pseudomonas aeruginosa PAO1', 'Pseudomonas_aeruginosa', 'Pseudomonas'] \n",
      " {'Pseudomonas'}\n",
      "\u001b[0;1;31;1mdistinct species: \u001b[0;1;30;1mCounter({'Pseudomonas_aeruginosa': 171, 'Pseudomonas_chlororaphis': 42, 'Pseudomonas_syringae': 29, 'Pseudomonas_putida': 28, 'Pseudomonas_fluorescens': 21, 'Pseudomonas_stutzeri': 17, 'Pseudomonas_protegens': 6, 'Pseudomonas_synxantha': 6, 'Pseudomonas_mendocina': 6, 'Pseudomonas_monteilii': 5, 'Pseudomonas_orientalis': 5, 'Pseudomonas_brassicacearum': 4, 'Pseudomonas_koreensis': 3, 'Pseudomonas_entomophila': 3, 'Pseudomonas_frederiksbergensis': 3, 'Pseudomonas_parafulva': 3, 'Pseudomonas_fragi': 2, 'Pseudomonas_taetrolens': 2, 'Pseudomonas_savastanoi': 2, 'Pseudomonas_fulva': 2, 'Pseudomonas_azotoformans': 2, 'Pseudomonas_pseudoalcaligenes': 2, 'Pseudomonas_mosselii': 2, 'Pseudomonas_plecoglossicida': 2, 'Pseudomonas_amygdali': 2, 'Pseudomonas_cerasi': 2, 'Pseudomonas_citronellolis': 2, 'Pseudomonas_cichorii': 1, 'Pseudomonas_resinovorans': 1, 'Pseudomonas_oryzihabitans': 1, 'Pseudomonas_alkylphenolica': 1, 'Pseudomonas_alcaliphila': 1, 'Pseudomonas_alcaligenes': 1, 'Pseudomonas_balearica': 1, 'Pseudomonas_libanensis': 1, 'Pseudomonas_silesiensis': 1, 'Pseudomonas_mandelii': 1, 'Pseudomonas_corrugata': 1, 'Pseudomonas_veronii': 1, 'Pseudomonas_yamanorum': 1, 'Pseudomonas_antarctica': 1, 'Pseudomonas_soli': 1, 'Pseudomonas_furukawaii': 1, 'Pseudomonas_knackmussii': 1, 'Pseudomonas_poae': 1, 'Pseudomonas_versuta': 1, 'Pseudomonas_trivialis': 1, 'Pseudomonas_avellanae': 1, 'Pseudomonas_viridiflava': 1, 'Pseudomonas_cremoricolorata': 1, 'Pseudomonas_lurida': 1, 'Pseudomonas_kribbensis': 1, 'Pseudomonas_mucidolens': 1, 'Pseudomonas_rhizosphaerae': 1, 'Pseudomonas_psychrotolerans': 1})\u001b[0m\n",
      "\u001b[0;1;31;1m{'Pseudomonas_frederiksbergensis', 'Pseudomonas_protegens', 'Pseudomonas_monteilii', 'Pseudomonas_koreensis', 'Pseudomonas_mendocina', 'Pseudomonas_syringae', 'Pseudomonas_stutzeri', 'Pseudomonas_parafulva', 'Pseudomonas_fluorescens', 'Pseudomonas_aeruginosa', 'Pseudomonas_chlororaphis', 'Pseudomonas_putida', 'Pseudomonas_brassicacearum', 'Pseudomonas_entomophila', 'Pseudomonas_orientalis', 'Pseudomonas_synxantha'}\u001b[0;1;30;1m\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "## read species names translation tables and keep only genus+species \n",
    "file_name = glob.glob(\"/media/deolivl/QIB_deolivl/bigdata/\" + bacteria_genus + \"*/names.txt\")[0]\n",
    "file_lines = [line.strip() for line in open(file_name, 'r')]\n",
    "thistbl = [line.split('\\t') for line in file_lines[1:]] # concatenation of lists\n",
    "fnames = glob.glob(\"/media/deolivl/QIB_deolivl/bigdata/\" + bacteria_genus + \"*/GCF_*.gbff.gz\")\n",
    "idx=[i for j in thistbl for i,fname in enumerate(fnames) if j[0] in fname]\n",
    "table_filenames_species = [[fnames[idx[i]]] + j for i,j in enumerate(thistbl)]\n",
    "table_filenames_species = [line + [\"_\".join(line[2].split()[:2]).split('.')[0]] for line in table_filenames_species]\n",
    "table_filenames_species = [line + [line[2].split()[0]] for line in table_filenames_species]\n",
    "print_redblack (\"table_filenames_species[] has:\",\" filename,  GCF_id, full species_subsp name, species name (genus+sp), genus\\n\")\n",
    "\n",
    "print (table_filenames_species[0], \"\\n\", set([x[4] for x  in table_filenames_species]))\n",
    "\n",
    "species_counter = collections.Counter([x[3] for x in table_filenames_species if x[3][-3:] !=\"_sp\"])\n",
    "print_redblack (\"distinct species: \", species_counter)\n",
    "\n",
    "def return_species_name_from_table (table_fs):\n",
    "    return str(table_fs[3]) + \"_\" + str(re.search('GCF_(\\d+)\\.', table_fs[1]).group(1))\n",
    "\n",
    "tbl_fl_sp = [x  for x in table_filenames_species if species_counter[ x[3] ] > 2]\n",
    "print_redblack(set([x[3] for x  in tbl_fl_sp]))\n",
    "random.shuffle(tbl_fl_sp)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### main functions\n",
    "* find all operons, align sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def align_seqs (sequences=None, maxiters=12, infile=None, outfile=None, mafft = True):\n",
    "    print (\"started aligning...\", flush=True, end=\" \")\n",
    "    if (sequences is None) and (infile is None):\n",
    "        print (\"ERROR: You must give me an alignment object or file\")\n",
    "        return [] ## OTOH if both are present then infile is overwritten with contents of sequences[]\n",
    "    if infile is None:\n",
    "        ifl = \"/tmp/in.fas\"\n",
    "    else:\n",
    "        ifl = infile\n",
    "    if outfile is None:\n",
    "        ofl = \"/tmp/out.fas\"\n",
    "    else:\n",
    "        ofl = outfile\n",
    "    SeqIO.write(sequences, ifl, \"fasta\")\n",
    "    if (mafft is False):\n",
    "        proc_run = subprocess.check_output(\"muscle -in \" + ifl + \" -diags -maxiters \" + str(maxiters) + \" -out \" + ofl,\n",
    "                                       shell=True, universal_newlines=True)\n",
    "    else: # \"--parttree --6merpair\" and 0.123 is to avoid very long alignments\n",
    "        proc_run = subprocess.check_output(\"mafft --ep 0.3 --op 3.0 --auto \" + ifl + \" > \" + ofl,\n",
    "                                       shell=True, universal_newlines=True)\n",
    "      \n",
    "    aligned = AlignIO.read(ofl, \"fasta\")\n",
    "    print (\"Finished\",flush=True)\n",
    "    if infile is None:\n",
    "        os.system(\"rm -f \" + ifl)\n",
    "    if outfile is None:\n",
    "        os.system(\"rm -f \" + ofl)\n",
    "    return aligned\n",
    "\n",
    "def return_contig_locations (featurelist): # featurelist has list of [ [location1, location2], rnaname, strand ]\n",
    "    featurelist = sorted(featurelist, key = lambda x: int(x[0][0]))\n",
    "    rnablocks = []\n",
    "    rnabnames = []\n",
    "    rnasignal = []\n",
    "    tmp_loc = list(featurelist[0][0]) # start and end locations\n",
    "    tmp_nam = [featurelist[0][1]]    # gene name\n",
    "    tmp_sig = [featurelist[0][2]]\n",
    "    for feature in featurelist[1:]:\n",
    "        if (abs(tmp_loc[-1] - feature[0][0]) < 1000):\n",
    "            tmp_loc.extend(list(feature[0]))\n",
    "            tmp_nam.extend([feature[1]])\n",
    "            tmp_sig.extend([feature[2]])\n",
    "        else:\n",
    "            rnablocks.append(tmp_loc)\n",
    "            rnabnames.append(tmp_nam)\n",
    "            rnasignal.append(tmp_sig)\n",
    "            tmp_loc = list(feature[0])\n",
    "            tmp_nam = [feature[1]]\n",
    "            tmp_sig = [feature[2]]\n",
    "    rnablocks.append(tmp_loc)\n",
    "    rnabnames.append(tmp_nam)\n",
    "    rnasignal.append(tmp_sig)\n",
    "    return rnablocks, rnabnames, rnasignal\n",
    "    \n",
    "def get_rrna_from_genbank_to_list_and_dict (table_fs):\n",
    "    gbank = SeqIO.parse(gzip.open(table_fs[0], \"rt\"), \"genbank\")\n",
    "    list_of_features = []\n",
    "    list_of_seqs = []\n",
    "    genome = next(gbank) # no point in iterating over gbank, only this has whole information\n",
    "    for feature in genome.features:\n",
    "        if(feature.type == \"rRNA\"):\n",
    "            for rna in rrna_types:\n",
    "                if rna in feature.qualifiers['product'][0]:\n",
    "                    this_product = rna\n",
    "                    break\n",
    "            this_location = [int(i) for i in re.findall('\\d+',str(feature.location))]\n",
    "            this_strand = feature.strand\n",
    "            if 0 not in this_location:\n",
    "                list_of_features.append([this_location[:2], this_product, this_strand])\n",
    "    # now list_of_features is complete  for this genome record (gbank[0])\n",
    "    if not list_of_features:\n",
    "        print (\"no info from \", table_fs[0])\n",
    "        return None # some genomes have no relevant annotation\n",
    "    location, gnames, strand = return_contig_locations (list_of_features)\n",
    "    genome_len = len(genome.seq)\n",
    "    for i, (l,g,s) in enumerate(zip(location, gnames, strand)):\n",
    "        i_start = min(l); i_end = max(l)\n",
    "        if sum(s) < 0: ## hopefully they all have same signal\n",
    "            contig = genome.seq[i_start:i_end].reverse_complement()\n",
    "            seqname = \"s\" + \"\".join(g[::-1])\n",
    "        else:\n",
    "            contig = genome.seq[i_start:i_end]\n",
    "            seqname = \"s\" + \"\".join(g)\n",
    "        numbercode = '_{:02d}'.format(i)\n",
    "        seqname = return_species_name_from_table(table_fs) + numbercode\n",
    "        if (len(contig) > 3000) and (len(contig) < 8500): # some contigs have only 5S; some genomes have only this 5S annotated\n",
    "            list_of_seqs.append(SeqRecord(contig, id = seqname, description = seqname))\n",
    "    return list_of_seqs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### sample genomes and create unaligned list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-9-35f3c6b5fa8b>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      4\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mfile_counter\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m%\u001b[0m\u001b[0;36m20\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m         \u001b[0mprint\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfile_counter\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mend\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\" \"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mflush\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 6\u001b[0;31m     \u001b[0mrna_list\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mget_rrna_from_genbank_to_list_and_dict\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mtfs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      7\u001b[0m     \u001b[0;31m# (some list will have no usable features or b/c too long/too short; some dict has no usable features\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mrna_list\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-4-843c8a61bb50>\u001b[0m in \u001b[0;36mget_rrna_from_genbank_to_list_and_dict\u001b[0;34m(table_fs)\u001b[0m\n\u001b[1;32m     57\u001b[0m     \u001b[0mlist_of_features\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     58\u001b[0m     \u001b[0mlist_of_seqs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 59\u001b[0;31m     \u001b[0mgenome\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnext\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgbank\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;31m# no point in iterating over gbank, only this has whole information\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     60\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mfeature\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mgenome\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfeatures\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     61\u001b[0m         \u001b[0;32mif\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfeature\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtype\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"rRNA\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.5/dist-packages/Bio/SeqIO/__init__.py\u001b[0m in \u001b[0;36mparse\u001b[0;34m(handle, format, alphabet)\u001b[0m\n\u001b[1;32m    653\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Unknown format '%s'\"\u001b[0m \u001b[0;34m%\u001b[0m \u001b[0mformat\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    654\u001b[0m         \u001b[0;31m# This imposes some overhead... wait until we drop Python 2.4 to fix it\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 655\u001b[0;31m         \u001b[0;32mfor\u001b[0m \u001b[0mr\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    656\u001b[0m             \u001b[0;32myield\u001b[0m \u001b[0mr\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    657\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.5/dist-packages/Bio/GenBank/Scanner.py\u001b[0m in \u001b[0;36mparse_records\u001b[0;34m(self, handle, do_features)\u001b[0m\n\u001b[1;32m    487\u001b[0m         \u001b[0;31m# This is a generator function\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    488\u001b[0m         \u001b[0;32mwhile\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 489\u001b[0;31m             \u001b[0mrecord\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhandle\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdo_features\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    490\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mrecord\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    491\u001b[0m                 \u001b[0;32mbreak\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.5/dist-packages/Bio/GenBank/Scanner.py\u001b[0m in \u001b[0;36mparse\u001b[0;34m(self, handle, do_features)\u001b[0m\n\u001b[1;32m    471\u001b[0m                                     feature_cleaner=FeatureValueCleaner())\n\u001b[1;32m    472\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 473\u001b[0;31m         \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfeed\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhandle\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mconsumer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdo_features\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    474\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mconsumer\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    475\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.5/dist-packages/Bio/GenBank/Scanner.py\u001b[0m in \u001b[0;36mfeed\u001b[0;34m(self, handle, consumer, do_features)\u001b[0m\n\u001b[1;32m    443\u001b[0m         \u001b[0;31m# Features (common to both EMBL and GenBank):\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    444\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mdo_features\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 445\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_feed_feature_table\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mconsumer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparse_features\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mskip\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    446\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    447\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparse_features\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mskip\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m  \u001b[0;31m# ignore the data\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.5/dist-packages/Bio/GenBank/Scanner.py\u001b[0m in \u001b[0;36mparse_features\u001b[0;34m(self, skip)\u001b[0m\n\u001b[1;32m    211\u001b[0m                     \u001b[0mfeature_lines\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mline\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mFEATURE_QUALIFIER_INDENT\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstrip\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    212\u001b[0m                     \u001b[0mline\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhandle\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreadline\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 213\u001b[0;31m                 \u001b[0mfeatures\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparse_feature\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfeature_key\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfeature_lines\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    214\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mline\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mline\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    215\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mfeatures\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.5/dist-packages/Bio/GenBank/Scanner.py\u001b[0m in \u001b[0;36mparse_feature\u001b[0;34m(self, feature_key, lines)\u001b[0m\n\u001b[1;32m    324\u001b[0m                         \u001b[0mvalue\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'\\n'\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalue_list\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    325\u001b[0m                         \u001b[0;31m# DO NOT remove the quotes...\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 326\u001b[0;31m                         \u001b[0mqualifiers\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    327\u001b[0m                     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    328\u001b[0m                         \u001b[0;31m# Unquoted\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "good_fl_sp = []\n",
    "rna_operon = []  # includes all paralogs\n",
    "for file_counter, tfs in enumerate(tbl_fl_sp[:200]):\n",
    "    if not (file_counter+1)%20:\n",
    "        print (str(file_counter+1), end=\" \", flush=True)\n",
    "    rna_list = get_rrna_from_genbank_to_list_and_dict (tfs)\n",
    "    # (some list will have no usable features or b/c too long/too short; some dict has no usable features\n",
    "    if rna_list:\n",
    "        good_fl_sp.append(tfs)\n",
    "        rna_operon.extend(rna_list)\n",
    "\n",
    "# save table, remembering that 'outdir' has the genus name already\n",
    "fl=gzip.open(outdir+\"_fl_sp.pickle.gz\", \"w\"); pickle.dump([good_fl_sp],fl,2); fl.close()\n",
    "\n",
    "print (\"# first sequence of operon alignment:\\n\", rna_operon[0])\n",
    "x = [len(i.seq) for i in rna_operon]\n",
    "print (\"# percentiles:\", np.min(x), np.percentile(x, 5), np.mean(x), np.percentile(x, 95), np.max(x), len(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "started aligning... Finished\n"
     ]
    }
   ],
   "source": [
    "# align each of the dictionary values (lists of unaligned seqs)\n",
    "operon_aligned = align_seqs(sequences=rna_operon, outfile=outdir + \".fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_species_genus_labels (seqlist, have_paralogs = False): \n",
    "    species = ['_'.join(sequence.id.split('_')[:2]) for sequence in seqlist]\n",
    "    genus = [sequence.id.split('_')[0] for sequence in seqlist]\n",
    "    if have_paralogs:# assumes 'genus_species_code_number' names \n",
    "        sample = ['_'.join(sequence.id.split('_')[:3]) for sequence in seqlist]\n",
    "        return species, genus, sample\n",
    "    return species, genus "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### if resuming from previous analysis, run code below to read genome names table and alignments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read table\n",
    "fl=gzip.open(outdir+\"_fl_sp.pickle.gz\", \"r\"); good_fl_sp = pickle.load(fl); fl.close()\n",
    "operon_aligned = AlignIO.read(outdir + \".fasta\", \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "IQ-TREE multicore version 1.3.11.1 for Linux 64-bit built Feb  1 2016\n",
      "Copyright (c) 2011-2015 Nguyen Lam Tung, Olga Chernomor, Arndt von Haeseler and Bui Quang Minh.\n",
      "\n",
      "Host:    n114312.nbi.ac.uk (AVX2, FMA3, 15 GB RAM)\n",
      "Command: iqtree-omp -s ./014_results/Pseudomonas.fasta -nt 8 -m HKY+G\n",
      "Seed:    410298 (Using SPRNG - Scalable Parallel Random Number Generator)\n",
      "Time:    Mon Feb 25 09:45:50 2019\n",
      "Kernel:  AVX - 8 threads (8 CPU cores detected)\n",
      "\n",
      "Reading alignment file ./014_results/Pseudomonas.fasta ... Fasta format detected\n",
      "Alignment most likely contains DNA/RNA sequences\n",
      "Alignment has 923 sequences with 7793 columns and 1739 patterns (1117 informative sites)\n",
      "                                             Gap/Ambiguity  Composition  p-value\n",
      "   1  Pseudomonas_aeruginosa_000510305_00           33.81%    passed     99.30%\n",
      "   2  Pseudomonas_aeruginosa_000510305_01           33.81%    passed     98.49%\n",
      "   3  Pseudomonas_aeruginosa_000510305_02           33.83%    passed     98.59%\n",
      "   4  Pseudomonas_aeruginosa_000510305_03           33.83%    passed     98.40%\n",
      "   5  Pseudomonas_aeruginosa_003950015_00           33.84%    passed     98.06%\n",
      "   6  Pseudomonas_aeruginosa_003950015_01           33.83%    passed     98.43%\n",
      "   7  Pseudomonas_aeruginosa_003950015_02           33.84%    passed     98.57%\n",
      "   8  Pseudomonas_aeruginosa_003950015_03           33.83%    passed     97.94%\n",
      "   9  Pseudomonas_aeruginosa_001721765_00           33.83%    passed     98.00%\n",
      "  10  Pseudomonas_aeruginosa_001721765_01           33.81%    passed     97.59%\n",
      "  11  Pseudomonas_aeruginosa_001721765_02           33.81%    passed     98.30%\n",
      "  12  Pseudomonas_aeruginosa_001721765_03           33.84%    passed     97.82%\n",
      "  13  Pseudomonas_brassicacearum_001449085_00       29.90%    passed     96.58%\n",
      "  14  Pseudomonas_brassicacearum_001449085_01       32.98%    passed     99.84%\n",
      "  15  Pseudomonas_brassicacearum_001449085_02       32.76%    passed     99.59%\n",
      "  16  Pseudomonas_brassicacearum_001449085_03       32.98%    passed     99.84%\n",
      "  17  Pseudomonas_brassicacearum_001449085_04       32.98%    passed     99.82%\n",
      "  18  Pseudomonas_aeruginosa_002812885_00           33.83%    passed     97.93%\n",
      "  19  Pseudomonas_aeruginosa_002812885_01           33.83%    passed     97.93%\n",
      "  20  Pseudomonas_aeruginosa_002812885_02           33.83%    passed     97.55%\n",
      "  21  Pseudomonas_aeruginosa_002812885_03           33.83%    passed     97.85%\n",
      "  22  Pseudomonas_aeruginosa_001547955_00           33.81%    passed     96.84%\n",
      "  23  Pseudomonas_aeruginosa_001547955_01           33.81%    passed     98.20%\n",
      "  24  Pseudomonas_aeruginosa_001547955_02           33.83%    passed     98.31%\n",
      "  25  Pseudomonas_aeruginosa_001547955_03           33.81%    passed     97.69%\n",
      "  26  Pseudomonas_aeruginosa_000271985_00           33.81%    passed     98.96%\n",
      "  27  Pseudomonas_aeruginosa_000271985_01           33.81%    passed     98.86%\n",
      "  28  Pseudomonas_aeruginosa_000271985_02           33.81%    passed     98.27%\n",
      "  29  Pseudomonas_aeruginosa_000271985_03           33.84%    passed     96.74%\n",
      "  30  Pseudomonas_chlororaphis_003851905_00         30.82%    passed     92.19%\n",
      "  31  Pseudomonas_chlororaphis_003851905_01         33.72%    passed     96.01%\n",
      "  32  Pseudomonas_chlororaphis_003851905_02         33.72%    passed     96.01%\n",
      "  33  Pseudomonas_chlororaphis_003851905_03         33.72%    passed     96.01%\n",
      "  34  Pseudomonas_chlororaphis_003851905_04         33.72%    passed     96.11%\n",
      "  35  Pseudomonas_protegens_000012265_00            30.28%    passed     61.24%\n",
      "  36  Pseudomonas_protegens_000012265_01            33.22%    passed     81.44%\n",
      "  37  Pseudomonas_protegens_000012265_02            33.22%    passed     79.78%\n",
      "  38  Pseudomonas_protegens_000012265_03            33.22%    passed     83.01%\n",
      "  39  Pseudomonas_protegens_000012265_04            33.22%    passed     80.01%\n",
      "  40  Pseudomonas_stutzeri_900636845_00             33.50%    passed     91.82%\n",
      "  41  Pseudomonas_stutzeri_900636845_01             33.50%    passed     91.49%\n",
      "  42  Pseudomonas_stutzeri_900636845_02             33.50%    passed     91.82%\n",
      "  43  Pseudomonas_stutzeri_900636845_03             33.50%    passed     89.81%\n",
      "  44  Pseudomonas_chlororaphis_003850525_00         30.84%    passed     94.10%\n",
      "  45  Pseudomonas_chlororaphis_003850525_01         33.72%    passed     95.55%\n",
      "  46  Pseudomonas_chlororaphis_003850525_02         33.72%    passed     96.29%\n",
      "  47  Pseudomonas_chlororaphis_003850525_03         33.57%    passed     96.73%\n",
      "  48  Pseudomonas_chlororaphis_003850525_04         33.72%    passed     95.55%\n",
      "  49  Pseudomonas_mendocina_002442555_00            33.07%    passed     90.54%\n",
      "  50  Pseudomonas_mendocina_002442555_01            33.43%    passed     94.26%\n",
      "  51  Pseudomonas_mendocina_002442555_02            33.07%    passed     91.17%\n",
      "  52  Pseudomonas_mendocina_002442555_03            33.06%    passed     90.92%\n",
      "  53  Pseudomonas_monteilii_000510325_00            35.93%    passed     97.93%\n",
      "  54  Pseudomonas_monteilii_000510325_01            35.93%    passed     98.42%\n",
      "  55  Pseudomonas_monteilii_000510325_02            35.93%    passed     98.85%\n",
      "  56  Pseudomonas_monteilii_000510325_03            35.93%    passed     98.14%\n",
      "  57  Pseudomonas_monteilii_000510325_04            35.93%    passed     98.85%\n",
      "  58  Pseudomonas_monteilii_000510325_05            33.08%    passed     99.41%\n",
      "  59  Pseudomonas_chlororaphis_003851925_00         30.84%    passed     91.08%\n",
      "  60  Pseudomonas_chlororaphis_003851925_01         33.74%    passed     93.49%\n",
      "  61  Pseudomonas_chlororaphis_003851925_02         33.61%    passed     95.95%\n",
      "  62  Pseudomonas_chlororaphis_003851925_03         33.61%    passed     95.59%\n",
      "  63  Pseudomonas_chlororaphis_003851925_04         33.72%    passed     95.13%\n",
      "  64  Pseudomonas_aeruginosa_001516305_00           33.83%    passed     98.40%\n",
      "  65  Pseudomonas_aeruginosa_001516305_01           33.83%    passed     97.67%\n",
      "  66  Pseudomonas_aeruginosa_001516305_02           33.83%    passed     98.77%\n",
      "  67  Pseudomonas_aeruginosa_001516305_03           33.83%    passed     98.15%\n",
      "  68  Pseudomonas_stutzeri_900638035_00             33.22%    passed     98.90%\n",
      "  69  Pseudomonas_stutzeri_900638035_01             33.22%    passed     99.13%\n",
      "  70  Pseudomonas_stutzeri_900638035_02             33.22%    passed     99.13%\n",
      "  71  Pseudomonas_stutzeri_900638035_03             33.22%    passed     99.30%\n",
      "  72  Pseudomonas_aeruginosa_001594325_00           33.81%    passed     97.70%\n",
      "  73  Pseudomonas_aeruginosa_001594325_01           33.83%    passed     98.40%\n",
      "  74  Pseudomonas_aeruginosa_001594325_02           33.83%    passed     98.40%\n",
      "  75  Pseudomonas_aeruginosa_001594325_03           33.81%    passed     97.80%\n",
      "  76  Pseudomonas_putida_001767335_00               33.17%    passed     97.29%\n",
      "  77  Pseudomonas_putida_001767335_01               36.08%    passed     99.34%\n",
      "  78  Pseudomonas_putida_001767335_02               36.24%    passed     99.39%\n",
      "  79  Pseudomonas_putida_001767335_03               33.36%    passed     99.01%\n",
      "  80  Pseudomonas_putida_001767335_04               36.08%    passed     99.34%\n",
      "  81  Pseudomonas_putida_001767335_05               33.35%    passed     99.15%\n",
      "  82  Pseudomonas_chlororaphis_003851145_00         30.78%    passed     91.60%\n",
      "  83  Pseudomonas_chlororaphis_003851145_01         33.70%    passed     95.53%\n",
      "  84  Pseudomonas_chlororaphis_003851145_02         33.68%    passed     97.34%\n",
      "  85  Pseudomonas_chlororaphis_003851145_03         33.70%    passed     96.24%\n",
      "  86  Pseudomonas_chlororaphis_003851145_04         33.68%    passed     96.41%\n",
      "  87  Pseudomonas_fluorescens_900636825_00          27.14%    passed     66.35%\n",
      "  88  Pseudomonas_fluorescens_900636825_01          33.03%    passed     98.79%\n",
      "  89  Pseudomonas_fluorescens_900636825_02          32.49%    passed     97.97%\n",
      "  90  Pseudomonas_fluorescens_900636825_03          33.02%    passed     98.68%\n",
      "  91  Pseudomonas_fluorescens_900636825_04          32.84%    passed     97.23%\n",
      "  92  Pseudomonas_fluorescens_900636825_05          33.03%    passed     98.93%\n",
      "  93  Pseudomonas_putida_002736125_01               36.06%    passed     94.56%\n",
      "  94  Pseudomonas_putida_002736125_02               32.44%    passed     98.25%\n",
      "  95  Pseudomonas_putida_002736125_03               36.05%    passed     94.52%\n",
      "  96  Pseudomonas_putida_002736125_04               36.07%    passed     94.39%\n",
      "  97  Pseudomonas_putida_002736125_05               33.58%    passed     98.91%\n",
      "  98  Pseudomonas_stutzeri_000219605_00             33.50%    passed     92.86%\n",
      "  99  Pseudomonas_stutzeri_000219605_01             33.50%    passed     90.68%\n",
      " 100  Pseudomonas_stutzeri_000219605_02             33.50%    passed     95.17%\n",
      " 101  Pseudomonas_stutzeri_000219605_03             33.50%    passed     92.21%\n",
      " 102  Pseudomonas_aeruginosa_001045685_00           33.83%    passed     98.40%\n",
      " 103  Pseudomonas_aeruginosa_001045685_01           33.83%    passed     98.77%\n",
      " 104  Pseudomonas_aeruginosa_001045685_02           33.83%    passed     98.77%\n",
      " 105  Pseudomonas_aeruginosa_001045685_03           33.83%    passed     98.77%\n",
      " 106  Pseudomonas_syringae_000452705_00             31.87%    passed     99.99%\n",
      " 107  Pseudomonas_syringae_000452705_01             32.59%    passed     99.95%\n",
      " 108  Pseudomonas_syringae_000452705_02             31.86%    passed     99.73%\n",
      " 109  Pseudomonas_syringae_000452705_03             32.79%    passed     99.02%\n",
      " 110  Pseudomonas_syringae_000452705_04             29.90%    passed     98.03%\n",
      " 111  Pseudomonas_syringae_900235865_00             32.77%    passed     99.36%\n",
      " 112  Pseudomonas_syringae_900235865_01             32.57%    passed     99.71%\n",
      " 113  Pseudomonas_syringae_900235865_02             32.76%    passed     98.71%\n",
      " 114  Pseudomonas_syringae_900235865_03             32.77%    passed     99.36%\n",
      " 115  Pseudomonas_syringae_900235865_04             29.90%    passed     99.13%\n",
      " 116  Pseudomonas_putida_000016865_00               33.35%    passed     97.60%\n",
      " 117  Pseudomonas_putida_000016865_01               33.81%    passed     99.40%\n",
      " 118  Pseudomonas_putida_000016865_02               36.08%    passed     98.58%\n",
      " 119  Pseudomonas_putida_000016865_03               33.81%    passed     99.30%\n",
      " 120  Pseudomonas_putida_000016865_04               33.52%    passed     98.90%\n",
      " 121  Pseudomonas_putida_000016865_05               33.52%    passed     98.84%\n",
      " 122  Pseudomonas_stutzeri_000195105_00             33.50%    passed     93.81%\n",
      " 123  Pseudomonas_stutzeri_000195105_01             33.50%    passed     93.41%\n",
      " 124  Pseudomonas_stutzeri_000195105_02             33.50%    passed     93.41%\n",
      " 125  Pseudomonas_stutzeri_000195105_03             33.50%    passed     93.81%\n",
      " 126  Pseudomonas_fluorescens_900475215_00          35.21%    passed     60.21%\n",
      " 127  Pseudomonas_fluorescens_900475215_01          29.99%    passed     69.45%\n",
      " 128  Pseudomonas_fluorescens_900475215_02          32.88%    passed     82.68%\n",
      " 129  Pseudomonas_fluorescens_900475215_03          32.88%    passed     79.11%\n",
      " 130  Pseudomonas_fluorescens_900475215_04          32.88%    passed     84.49%\n",
      " 131  Pseudomonas_fluorescens_900475215_05          32.88%    passed     80.52%\n",
      " 132  Pseudomonas_putida_001886975_01               36.35%    passed     99.85%\n",
      " 133  Pseudomonas_putida_001886975_02               33.40%    passed     99.32%\n",
      " 134  Pseudomonas_putida_001886975_03               33.40%    passed     99.40%\n",
      " 135  Pseudomonas_putida_001886975_04               36.35%    passed     99.79%\n",
      " 136  Pseudomonas_putida_001886975_05               36.35%    passed     99.91%\n",
      " 137  Pseudomonas_aeruginosa_003812905_00           33.86%    passed     99.17%\n",
      " 138  Pseudomonas_aeruginosa_003812905_01           33.86%    passed     98.70%\n",
      " 139  Pseudomonas_aeruginosa_003812905_02           33.84%    passed     99.04%\n",
      " 140  Pseudomonas_aeruginosa_003812905_03           33.85%    passed     99.11%\n",
      " 141  Pseudomonas_stutzeri_003008495_00             33.53%    passed     93.30%\n",
      " 142  Pseudomonas_stutzeri_003008495_01             33.53%    passed     93.18%\n",
      " 143  Pseudomonas_stutzeri_003008495_02             33.53%    passed     93.30%\n",
      " 144  Pseudomonas_stutzeri_003008495_03             33.53%    passed     93.18%\n",
      " 145  Pseudomonas_aeruginosa_002192475_00           33.83%    passed     98.55%\n",
      " 146  Pseudomonas_aeruginosa_002192475_01           33.83%    passed     98.40%\n",
      " 147  Pseudomonas_aeruginosa_002192475_02           33.84%    passed     98.91%\n",
      " 148  Pseudomonas_aeruginosa_002192475_03           33.83%    passed     98.40%\n",
      " 149  Pseudomonas_aeruginosa_002075065_00           33.83%    passed     99.24%\n",
      " 150  Pseudomonas_aeruginosa_002075065_01           33.83%    passed     99.13%\n",
      " 151  Pseudomonas_aeruginosa_002075065_02           33.83%    passed     99.45%\n",
      " 152  Pseudomonas_aeruginosa_002075065_03           33.83%    passed     99.24%\n",
      " 153  Pseudomonas_stutzeri_000279165_00             33.03%    passed     72.58%\n",
      " 154  Pseudomonas_stutzeri_000279165_01             32.97%    passed     69.10%\n",
      " 155  Pseudomonas_stutzeri_000279165_02             32.79%    passed     76.99%\n",
      " 156  Pseudomonas_stutzeri_000279165_03             32.82%    passed     59.21%\n",
      " 157  Pseudomonas_aeruginosa_001515845_00           33.83%    passed     98.15%\n",
      " 158  Pseudomonas_aeruginosa_001515845_01           33.83%    passed     97.94%\n",
      " 159  Pseudomonas_aeruginosa_001515845_02           33.81%    passed     98.71%\n",
      " 160  Pseudomonas_aeruginosa_001515845_03           59.48%    passed     87.79%\n",
      " 161  Pseudomonas_chlororaphis_003850585_00         30.84%    passed     94.10%\n",
      " 162  Pseudomonas_chlororaphis_003850585_01         33.72%    passed     96.29%\n",
      " 163  Pseudomonas_chlororaphis_003850585_02         33.72%    passed     95.57%\n",
      " 164  Pseudomonas_chlororaphis_003850585_03         33.71%    passed     97.31%\n",
      " 165  Pseudomonas_chlororaphis_003850585_04         33.71%    passed     97.31%\n",
      " 166  Pseudomonas_aeruginosa_003288435_00           33.89%    passed     98.22%\n",
      " 167  Pseudomonas_aeruginosa_003288435_01           33.84%    passed     98.51%\n",
      " 168  Pseudomonas_aeruginosa_003288435_02           33.84%    passed     98.68%\n",
      " 169  Pseudomonas_aeruginosa_003288435_03           33.85%    passed     98.61%\n",
      " 170  Pseudomonas_chlororaphis_003850465_00         30.84%    passed     94.10%\n",
      " 171  Pseudomonas_chlororaphis_003850465_01         33.72%    passed     94.77%\n",
      " 172  Pseudomonas_chlororaphis_003850465_02         33.72%    passed     95.55%\n",
      " 173  Pseudomonas_chlororaphis_003850465_03         33.72%    passed     94.77%\n",
      " 174  Pseudomonas_chlororaphis_003850465_04         33.57%    passed     96.73%\n",
      " 175  Pseudomonas_chlororaphis_003008635_00         30.82%    passed     92.28%\n",
      " 176  Pseudomonas_chlororaphis_003008635_01         33.72%    passed     95.88%\n",
      " 177  Pseudomonas_chlororaphis_003008635_02         33.72%    passed     95.88%\n",
      " 178  Pseudomonas_chlororaphis_003008635_03         33.72%    passed     95.88%\n",
      " 179  Pseudomonas_chlororaphis_003008635_04         33.72%    passed     96.92%\n",
      " 180  Pseudomonas_aeruginosa_001516325_00           33.83%    passed     97.94%\n",
      " 181  Pseudomonas_aeruginosa_001516325_01           33.83%    passed     96.44%\n",
      " 182  Pseudomonas_aeruginosa_001516325_02           33.83%    passed     98.40%\n",
      " 183  Pseudomonas_aeruginosa_001516325_03           33.81%    passed     97.51%\n",
      " 184  Pseudomonas_stutzeri_000013785_00             33.50%    passed     94.85%\n",
      " 185  Pseudomonas_stutzeri_000013785_01             33.50%    passed     96.82%\n",
      " 186  Pseudomonas_stutzeri_000013785_02             33.50%    passed     93.60%\n",
      " 187  Pseudomonas_stutzeri_000013785_03             32.93%    passed     99.12%\n",
      " 188  Pseudomonas_syringae_004006335_00             30.16%    passed     96.65%\n",
      " 189  Pseudomonas_syringae_004006335_01             33.17%    passed     97.58%\n",
      " 190  Pseudomonas_syringae_004006335_02             33.00%    passed     99.14%\n",
      " 191  Pseudomonas_syringae_004006335_03             33.17%    passed     98.64%\n",
      " 192  Pseudomonas_syringae_004006335_04             33.18%    passed     98.61%\n",
      " 193  Pseudomonas_syringae_004006335_05             33.18%    passed     98.61%\n",
      " 194  Pseudomonas_syringae_001482725_00             32.76%    passed     99.32%\n",
      " 195  Pseudomonas_syringae_001482725_01             32.57%    passed     99.89%\n",
      " 196  Pseudomonas_syringae_001482725_02             32.57%    passed     99.88%\n",
      " 197  Pseudomonas_syringae_001482725_03             32.57%    passed     99.89%\n",
      " 198  Pseudomonas_syringae_001482725_04             29.71%    passed     97.20%\n",
      " 199  Pseudomonas_aeruginosa_002812925_00           33.84%    passed     99.04%\n",
      " 200  Pseudomonas_aeruginosa_002812925_01           33.84%    passed     99.04%\n",
      " 201  Pseudomonas_aeruginosa_002812925_02           33.85%    passed     98.61%\n",
      " 202  Pseudomonas_aeruginosa_002812925_03           33.84%    passed     99.04%\n",
      " 203  Pseudomonas_aeruginosa_000496455_00           33.83%    passed     98.59%\n",
      " 204  Pseudomonas_aeruginosa_000496455_01           33.83%    passed     98.59%\n",
      " 205  Pseudomonas_aeruginosa_000496455_02           33.81%    passed     98.86%\n",
      " 206  Pseudomonas_aeruginosa_000496455_03           33.83%    passed     98.77%\n",
      " 207  Pseudomonas_fluorescens_003410335_00          33.15%    passed     99.13%\n",
      " 208  Pseudomonas_fluorescens_003410335_01          32.93%    passed     98.14%\n",
      " 209  Pseudomonas_fluorescens_003410335_02          32.94%    passed     98.11%\n",
      " 210  Pseudomonas_fluorescens_003410335_03          30.09%    passed     98.59%\n",
      " 211  Pseudomonas_fluorescens_003410335_04          32.93%    passed     98.61%\n",
      " 212  Pseudomonas_fluorescens_000963495_00          32.57%    passed     70.58%\n",
      " 213  Pseudomonas_fluorescens_000963495_01          30.19%    passed     62.57%\n",
      " 214  Pseudomonas_fluorescens_000963495_02          32.57%    passed     70.58%\n",
      " 215  Pseudomonas_fluorescens_000963495_03          33.08%    passed     77.47%\n",
      " 216  Pseudomonas_fluorescens_000963495_04          35.66%    passed     60.99%\n",
      " 217  Pseudomonas_fluorescens_000963495_05          32.57%    passed     72.43%\n",
      " 218  Pseudomonas_syringae_000988485_00             32.55%    passed     99.97%\n",
      " 219  Pseudomonas_syringae_000988485_01             32.55%    passed     99.96%\n",
      " 220  Pseudomonas_syringae_000988485_02             32.76%    passed     99.19%\n",
      " 221  Pseudomonas_syringae_000988485_03             32.76%    passed     99.32%\n",
      " 222  Pseudomonas_syringae_000988485_04             29.90%    passed     99.02%\n",
      " 223  Pseudomonas_chlororaphis_003851405_00         30.69%    passed     91.95%\n",
      " 224  Pseudomonas_chlororaphis_003851405_01         33.59%    passed     96.92%\n",
      " 225  Pseudomonas_chlororaphis_003851405_02         33.59%    passed     96.92%\n",
      " 226  Pseudomonas_chlororaphis_003851405_03         33.59%    passed     97.06%\n",
      " 227  Pseudomonas_chlororaphis_003851405_04         33.59%    passed     96.92%\n",
      " 228  Pseudomonas_aeruginosa_000359505_00           33.81%    passed     97.91%\n",
      " 229  Pseudomonas_aeruginosa_000359505_01           33.81%    passed     98.22%\n",
      " 230  Pseudomonas_aeruginosa_000359505_02           33.81%    passed     98.36%\n",
      " 231  Pseudomonas_aeruginosa_000359505_03           33.81%    passed     98.03%\n",
      " 232  Pseudomonas_stutzeri_000327065_00             33.25%    passed     98.63%\n",
      " 233  Pseudomonas_stutzeri_000327065_01             33.21%    passed     98.81%\n",
      " 234  Pseudomonas_stutzeri_000327065_02             33.22%    passed     98.64%\n",
      " 235  Pseudomonas_stutzeri_000327065_03             33.22%    passed     98.62%\n",
      " 236  Pseudomonas_aeruginosa_900618315_00           33.84%    passed     98.29%\n",
      " 237  Pseudomonas_aeruginosa_900618315_01           33.83%    passed     98.43%\n",
      " 238  Pseudomonas_aeruginosa_900618315_02           33.84%    passed     98.29%\n",
      " 239  Pseudomonas_aeruginosa_900618315_03           33.84%    passed     98.68%\n",
      " 240  Pseudomonas_aeruginosa_003812885_00           33.84%    passed     99.04%\n",
      " 241  Pseudomonas_aeruginosa_003812885_01           33.84%    passed     98.51%\n",
      " 242  Pseudomonas_aeruginosa_003812885_02           33.84%    passed     99.04%\n",
      " 243  Pseudomonas_aeruginosa_003812885_03           33.85%    passed     99.11%\n",
      " 244  Pseudomonas_aeruginosa_000284555_00           33.84%    passed     99.04%\n",
      " 245  Pseudomonas_aeruginosa_000284555_01           33.84%    passed     99.04%\n",
      " 246  Pseudomonas_aeruginosa_000284555_02           33.84%    passed     98.51%\n",
      " 247  Pseudomonas_aeruginosa_000284555_03           33.84%    passed     99.04%\n",
      " 248  Pseudomonas_fluorescens_000012445_00          30.21%    passed     97.42%\n",
      " 249  Pseudomonas_fluorescens_000012445_01          33.04%    passed     99.39%\n",
      " 250  Pseudomonas_fluorescens_000012445_02          33.04%    passed     99.82%\n",
      " 251  Pseudomonas_fluorescens_000012445_03          33.06%    passed     99.93%\n",
      " 252  Pseudomonas_fluorescens_000012445_04          33.17%    passed     99.67%\n",
      " 253  Pseudomonas_fluorescens_000012445_05          33.04%    passed     99.39%\n",
      " 254  Pseudomonas_chlororaphis_002844145_00         30.69%    passed     92.14%\n",
      " 255  Pseudomonas_chlororaphis_002844145_01         33.47%    passed     89.74%\n",
      " 256  Pseudomonas_chlororaphis_002844145_02         33.59%    passed     96.99%\n",
      " 257  Pseudomonas_chlororaphis_002844145_03         34.13%    passed     72.90%\n",
      " 258  Pseudomonas_chlororaphis_002844145_04         33.59%    passed     96.76%\n",
      " 259  Pseudomonas_brassicacearum_000585995_00       33.12%    passed     99.35%\n",
      " 260  Pseudomonas_brassicacearum_000585995_01       32.97%    passed     99.79%\n",
      " 261  Pseudomonas_brassicacearum_000585995_02       32.71%    passed     99.55%\n",
      " 262  Pseudomonas_brassicacearum_000585995_03       32.71%    passed     99.87%\n",
      " 263  Pseudomonas_brassicacearum_000585995_04       29.83%    passed     95.31%\n",
      " 264  Pseudomonas_entomophila_003940785_00          30.59%    passed     95.76%\n",
      " 265  Pseudomonas_entomophila_003940785_01          33.48%    passed     99.28%\n",
      " 266  Pseudomonas_entomophila_003940785_02          33.48%    passed     99.27%\n",
      " 267  Pseudomonas_entomophila_003940785_03          33.40%    passed     99.34%\n",
      " 268  Pseudomonas_entomophila_003940785_04          33.48%    passed     99.43%\n",
      " 269  Pseudomonas_entomophila_003940785_05          35.97%    passed     95.05%\n",
      " 270  Pseudomonas_entomophila_003940785_06          33.43%    passed     99.57%\n",
      " 271  Pseudomonas_chlororaphis_003851225_00         30.84%    passed     92.47%\n",
      " 272  Pseudomonas_chlororaphis_003851225_01         33.74%    passed     95.75%\n",
      " 273  Pseudomonas_chlororaphis_003851225_02         33.74%    passed     95.75%\n",
      " 274  Pseudomonas_chlororaphis_003851225_03         33.74%    passed     96.58%\n",
      " 275  Pseudomonas_chlororaphis_003851225_04         33.74%    passed     96.58%\n",
      " 276  Pseudomonas_putida_003671955_01               33.95%    passed     99.88%\n",
      " 277  Pseudomonas_putida_003671955_02               33.95%    passed     99.87%\n",
      " 278  Pseudomonas_putida_003671955_03               33.94%    passed     99.85%\n",
      " 279  Pseudomonas_putida_003671955_04               33.95%    passed     99.79%\n",
      " 280  Pseudomonas_putida_003671955_05               33.95%    passed     99.87%\n",
      " 281  Pseudomonas_putida_000325725_01               36.35%    passed     99.87%\n",
      " 282  Pseudomonas_putida_000325725_02               36.08%    passed     98.58%\n",
      " 283  Pseudomonas_putida_000325725_03               33.36%    passed     98.89%\n",
      " 284  Pseudomonas_putida_000325725_04               36.35%    passed     99.92%\n",
      " 285  Pseudomonas_putida_000325725_05               32.53%    passed     96.57%\n",
      " 286  Pseudomonas_aeruginosa_001722025_00           33.83%    passed     98.00%\n",
      " 287  Pseudomonas_aeruginosa_001722025_01           33.81%    passed     97.59%\n",
      " 288  Pseudomonas_aeruginosa_001722025_02           33.81%    passed     98.30%\n",
      " 289  Pseudomonas_aeruginosa_001722025_03           33.84%    passed     97.82%\n",
      " 290  Pseudomonas_synxantha_000968415_00            33.08%    passed     93.19%\n",
      " 291  Pseudomonas_synxantha_000968415_01            33.08%    passed     93.19%\n",
      " 292  Pseudomonas_synxantha_000968415_02            33.09%    passed     92.99%\n",
      " 293  Pseudomonas_synxantha_000968415_03            33.09%    passed     92.64%\n",
      " 294  Pseudomonas_synxantha_000968415_04            30.08%    passed     74.18%\n",
      " 295  Pseudomonas_synxantha_000968415_05            33.09%    passed     89.23%\n",
      " 296  Pseudomonas_aeruginosa_003571505_00           33.81%    passed     99.38%\n",
      " 297  Pseudomonas_aeruginosa_003571505_01           33.81%    passed     99.15%\n",
      " 298  Pseudomonas_aeruginosa_003571505_02           33.81%    passed     99.38%\n",
      " 299  Pseudomonas_aeruginosa_003571505_03           33.83%    passed     99.31%\n",
      " 300  Pseudomonas_aeruginosa_001900265_00           33.81%    passed     98.49%\n",
      " 301  Pseudomonas_aeruginosa_001900265_03           33.81%    passed     98.49%\n",
      " 302  Pseudomonas_aeruginosa_001900265_04           33.81%    passed     98.49%\n",
      " 303  Pseudomonas_aeruginosa_001792875_00           33.81%    passed     98.27%\n",
      " 304  Pseudomonas_aeruginosa_001792875_01           33.81%    passed     98.27%\n",
      " 305  Pseudomonas_aeruginosa_001792875_02           33.83%    passed     98.15%\n",
      " 306  Pseudomonas_aeruginosa_001792875_03           33.83%    passed     98.15%\n",
      " 307  Pseudomonas_chlororaphis_003851205_00         30.82%    passed     94.15%\n",
      " 308  Pseudomonas_chlororaphis_003851205_01         33.72%    passed     96.65%\n",
      " 309  Pseudomonas_chlororaphis_003851205_02         33.72%    passed     96.65%\n",
      " 310  Pseudomonas_chlororaphis_003851205_03         33.72%    passed     96.65%\n",
      " 311  Pseudomonas_chlororaphis_003851205_04         33.72%    passed     96.79%\n",
      " 312  Pseudomonas_aeruginosa_002946935_00           33.81%    passed     96.93%\n",
      " 313  Pseudomonas_aeruginosa_002946935_01           33.83%    passed     96.73%\n",
      " 314  Pseudomonas_aeruginosa_002946935_02           33.83%    passed     97.67%\n",
      " 315  Pseudomonas_aeruginosa_002946935_03           33.81%    passed     96.93%\n",
      " 316  Pseudomonas_aeruginosa_001792855_00           33.81%    passed     98.27%\n",
      " 317  Pseudomonas_aeruginosa_001792855_01           33.81%    passed     98.27%\n",
      " 318  Pseudomonas_aeruginosa_001792855_02           33.83%    passed     98.15%\n",
      " 319  Pseudomonas_aeruginosa_001792855_03           33.83%    passed     98.15%\n",
      " 320  Pseudomonas_protegens_002006545_00            30.48%    passed     56.98%\n",
      " 321  Pseudomonas_protegens_002006545_01            33.29%    passed     79.96%\n",
      " 322  Pseudomonas_protegens_002006545_02            33.31%    passed     78.79%\n",
      " 323  Pseudomonas_protegens_002006545_03            33.30%    passed     79.49%\n",
      " 324  Pseudomonas_protegens_002006545_04            33.31%    passed     78.86%\n",
      " 325  Pseudomonas_putida_003290365_00               34.16%    passed     97.55%\n",
      " 326  Pseudomonas_putida_003290365_01               32.82%    passed     97.73%\n",
      " 327  Pseudomonas_putida_003290365_02               32.82%    passed     97.69%\n",
      " 328  Pseudomonas_putida_003290365_03               34.09%    passed     96.28%\n",
      " 329  Pseudomonas_putida_003290365_04               34.15%    passed     98.74%\n",
      " 330  Pseudomonas_putida_003290365_05               31.39%    passed     93.62%\n",
      " 331  Pseudomonas_chlororaphis_000698865_00         56.65%    passed     81.46%\n",
      " 332  Pseudomonas_chlororaphis_000698865_01         33.72%    passed     94.40%\n",
      " 333  Pseudomonas_chlororaphis_000698865_02         33.72%    passed     93.14%\n",
      " 334  Pseudomonas_chlororaphis_000698865_03         33.72%    passed     94.21%\n",
      " 335  Pseudomonas_chlororaphis_000698865_04         33.72%    passed     94.21%\n",
      " 336  Pseudomonas_protegens_003363755_00            33.36%    passed     86.16%\n",
      " 337  Pseudomonas_protegens_003363755_01            33.36%    passed     86.99%\n",
      " 338  Pseudomonas_protegens_003363755_02            30.42%    passed     73.73%\n",
      " 339  Pseudomonas_protegens_003363755_03            30.42%    passed     73.22%\n",
      " 340  Pseudomonas_protegens_003363755_04            33.36%    passed     86.99%\n",
      " 341  Pseudomonas_aeruginosa_002237425_00           33.84%    passed     97.53%\n",
      " 342  Pseudomonas_aeruginosa_002237425_01           33.84%    passed     97.53%\n",
      " 343  Pseudomonas_aeruginosa_002237425_02           33.84%    passed     97.64%\n",
      " 344  Pseudomonas_aeruginosa_002237425_03           33.84%    passed     98.29%\n",
      " 345  Pseudomonas_aeruginosa_002753655_00           33.84%    passed     96.81%\n",
      " 346  Pseudomonas_aeruginosa_002753655_01           33.85%    passed     98.02%\n",
      " 347  Pseudomonas_aeruginosa_002753655_02           33.85%    passed     99.11%\n",
      " 348  Pseudomonas_aeruginosa_002753655_03           33.85%    passed     98.31%\n",
      " 349  Pseudomonas_stutzeri_003952685_00             32.73%    passed     85.62%\n",
      " 350  Pseudomonas_stutzeri_003952685_01             32.91%    passed     92.63%\n",
      " 351  Pseudomonas_stutzeri_003952685_02             30.05%    passed     83.21%\n",
      " 352  Pseudomonas_stutzeri_003952685_03             32.91%    passed     93.36%\n",
      " 353  Pseudomonas_stutzeri_003952685_04             35.45%    passed     80.78%\n",
      " 354  Pseudomonas_stutzeri_003952685_05             32.75%    passed     83.90%\n",
      " 355  Pseudomonas_stutzeri_003952685_06             32.91%    passed     92.63%\n",
      " 356  Pseudomonas_chlororaphis_003851865_00         30.82%    passed     93.49%\n",
      " 357  Pseudomonas_chlororaphis_003851865_01         33.72%    passed     96.11%\n",
      " 358  Pseudomonas_chlororaphis_003851865_02         33.74%    passed     94.55%\n",
      " 359  Pseudomonas_chlororaphis_003851865_03         33.72%    passed     96.01%\n",
      " 360  Pseudomonas_chlororaphis_003851865_04         33.72%    passed     96.11%\n",
      " 361  Pseudomonas_chlororaphis_003851265_00         30.73%    passed     95.01%\n",
      " 362  Pseudomonas_chlororaphis_003851265_01         33.75%    passed     97.44%\n",
      " 363  Pseudomonas_chlororaphis_003851265_02         33.63%    passed     99.07%\n",
      " 364  Pseudomonas_chlororaphis_003851265_03         33.74%    passed     97.37%\n",
      " 365  Pseudomonas_chlororaphis_003851265_04         33.74%    passed     97.98%\n",
      " 366  Pseudomonas_mendocina_000016565_00            33.06%    passed     94.58%\n",
      " 367  Pseudomonas_mendocina_000016565_01            33.06%    passed     93.61%\n",
      " 368  Pseudomonas_mendocina_000016565_02            33.06%    passed     93.61%\n",
      " 369  Pseudomonas_mendocina_000016565_03            33.07%    passed     93.17%\n",
      " 370  Pseudomonas_protegens_000828695_00            30.28%    passed     54.91%\n",
      " 371  Pseudomonas_protegens_000828695_01            33.22%    passed     82.39%\n",
      " 372  Pseudomonas_protegens_000828695_02            33.21%    passed     81.95%\n",
      " 373  Pseudomonas_protegens_000828695_03            33.21%    passed     81.95%\n",
      " 374  Pseudomonas_protegens_000828695_04            33.21%    passed     81.95%\n",
      " 375  Pseudomonas_aeruginosa_001516165_00           33.81%    passed     98.66%\n",
      " 376  Pseudomonas_aeruginosa_001516165_01           33.83%    passed     98.59%\n",
      " 377  Pseudomonas_aeruginosa_001516165_02           33.84%    passed     98.69%\n",
      " 378  Pseudomonas_aeruginosa_001516165_03           33.81%    passed     98.66%\n",
      " 379  Pseudomonas_fluorescens_001307275_00          32.72%    passed     99.80%\n",
      " 380  Pseudomonas_fluorescens_001307275_01          29.86%    passed     99.02%\n",
      " 381  Pseudomonas_fluorescens_001307275_02          32.97%    passed     99.19%\n",
      " 382  Pseudomonas_fluorescens_001307275_03          32.73%    passed     99.69%\n",
      " 383  Pseudomonas_fluorescens_001307275_04          32.97%    passed     99.37%\n",
      " 384  Pseudomonas_aeruginosa_000496605_00           33.81%    passed     97.51%\n",
      " 385  Pseudomonas_aeruginosa_000496605_01           33.81%    passed     97.51%\n",
      " 386  Pseudomonas_aeruginosa_000496605_02           33.81%    passed     96.93%\n",
      " 387  Pseudomonas_aeruginosa_000496605_03           33.81%    passed     96.93%\n",
      " 388  Pseudomonas_aeruginosa_000504045_00           33.83%    passed     97.94%\n",
      " 389  Pseudomonas_aeruginosa_000504045_01           33.83%    passed     97.94%\n",
      " 390  Pseudomonas_aeruginosa_000504045_02           33.81%    passed     98.49%\n",
      " 391  Pseudomonas_aeruginosa_000504045_03           33.83%    passed     97.94%\n",
      " 392  Pseudomonas_chlororaphis_002355875_00         33.72%    passed     95.13%\n",
      " 393  Pseudomonas_chlororaphis_002355875_01         33.70%    passed     96.00%\n",
      " 394  Pseudomonas_chlororaphis_002355875_02         33.74%    passed     95.12%\n",
      " 395  Pseudomonas_chlororaphis_002355875_03         33.74%    passed     95.12%\n",
      " 396  Pseudomonas_chlororaphis_002355875_04         30.84%    passed     91.88%\n",
      " 397  Pseudomonas_aeruginosa_001516205_00           33.83%    passed     97.94%\n",
      " 398  Pseudomonas_aeruginosa_001516205_01           33.83%    passed     97.35%\n",
      " 399  Pseudomonas_aeruginosa_001516205_02           33.81%    passed     99.02%\n",
      " 400  Pseudomonas_aeruginosa_001516205_03           33.83%    passed     97.94%\n",
      " 401  Pseudomonas_aeruginosa_003206535_00           33.83%    passed     98.15%\n",
      " 402  Pseudomonas_aeruginosa_003206535_01           33.83%    passed     97.67%\n",
      " 403  Pseudomonas_aeruginosa_003206535_02           33.84%    passed     98.38%\n",
      " 404  Pseudomonas_aeruginosa_003206535_03           33.83%    passed     98.25%\n",
      " 405  Pseudomonas_aeruginosa_002085755_00           33.84%    passed     99.04%\n",
      " 406  Pseudomonas_aeruginosa_002085755_01           33.84%    passed     99.04%\n",
      " 407  Pseudomonas_aeruginosa_002085755_02           33.83%    passed     98.59%\n",
      " 408  Pseudomonas_aeruginosa_002085755_03           33.84%    passed     99.04%\n",
      " 409  Pseudomonas_chlororaphis_003851985_00         30.78%    passed     94.98%\n",
      " 410  Pseudomonas_chlororaphis_003851985_01         33.68%    passed     96.12%\n",
      " 411  Pseudomonas_chlororaphis_003851985_02         33.68%    passed     96.12%\n",
      " 412  Pseudomonas_chlororaphis_003851985_03         33.68%    passed     96.12%\n",
      " 413  Pseudomonas_chlororaphis_003851985_04         33.68%    passed     96.12%\n",
      " 414  Pseudomonas_aeruginosa_900636735_00           33.83%    passed     98.96%\n",
      " 415  Pseudomonas_aeruginosa_900636735_01           33.83%    passed     98.96%\n",
      " 416  Pseudomonas_aeruginosa_900636735_02           33.83%    passed     97.94%\n",
      " 417  Pseudomonas_aeruginosa_900636735_03           33.83%    passed     98.96%\n",
      " 418  Pseudomonas_putida_000495455_00               33.36%    passed     98.99%\n",
      " 419  Pseudomonas_putida_000495455_01               33.36%    passed     98.55%\n",
      " 420  Pseudomonas_putida_000495455_02               33.36%    passed     98.79%\n",
      " 421  Pseudomonas_putida_000495455_03               33.36%    passed     98.99%\n",
      " 422  Pseudomonas_putida_000495455_04               33.47%    passed     98.95%\n",
      " 423  Pseudomonas_putida_000495455_05               33.62%    passed     99.19%\n",
      " 424  Pseudomonas_aeruginosa_003813005_00           33.88%    passed     98.93%\n",
      " 425  Pseudomonas_aeruginosa_003813005_01           33.92%    passed     98.72%\n",
      " 426  Pseudomonas_aeruginosa_003813005_02           33.86%    passed     98.52%\n",
      " 427  Pseudomonas_aeruginosa_003813005_03           33.90%    passed     98.48%\n",
      " 428  Pseudomonas_putida_000264665_00               33.50%    passed     99.24%\n",
      " 429  Pseudomonas_putida_000264665_01               33.50%    passed     99.26%\n",
      " 430  Pseudomonas_putida_000264665_02               33.21%    passed     97.46%\n",
      " 431  Pseudomonas_putida_000264665_03               33.97%    passed     99.69%\n",
      " 432  Pseudomonas_putida_000264665_04               36.08%    passed     98.60%\n",
      " 433  Pseudomonas_putida_000264665_05               33.81%    passed     99.59%\n",
      " 434  Pseudomonas_fluorescens_003055645_00          29.78%    passed     96.59%\n",
      " 435  Pseudomonas_fluorescens_003055645_01          33.17%    passed     99.56%\n",
      " 436  Pseudomonas_fluorescens_003055645_02          32.94%    passed     97.02%\n",
      " 437  Pseudomonas_fluorescens_003055645_03          32.62%    passed     99.29%\n",
      " 438  Pseudomonas_fluorescens_003055645_04          32.94%    passed     96.59%\n",
      " 439  Pseudomonas_fluorescens_003055645_05          33.50%    passed     98.27%\n",
      " 440  Pseudomonas_chlororaphis_003850505_00         30.84%    passed     94.10%\n",
      " 441  Pseudomonas_chlororaphis_003850505_01         33.72%    passed     95.55%\n",
      " 442  Pseudomonas_chlororaphis_003850505_02         33.72%    passed     96.29%\n",
      " 443  Pseudomonas_chlororaphis_003850505_03         33.71%    passed     97.31%\n",
      " 444  Pseudomonas_chlororaphis_003850505_04         33.57%    passed     96.73%\n",
      " 445  Pseudomonas_aeruginosa_003073895_00           33.81%    passed     98.87%\n",
      " 446  Pseudomonas_aeruginosa_003073895_01           33.81%    passed     99.04%\n",
      " 447  Pseudomonas_aeruginosa_003073895_02           33.84%    passed     97.53%\n",
      " 448  Pseudomonas_aeruginosa_003073895_03           33.83%    passed     98.40%\n",
      " 449  Pseudomonas_aeruginosa_003025345_00           33.85%    passed     98.04%\n",
      " 450  Pseudomonas_aeruginosa_003025345_01           33.85%    passed     97.75%\n",
      " 451  Pseudomonas_aeruginosa_003025345_02           33.85%    passed     97.75%\n",
      " 452  Pseudomonas_aeruginosa_003025345_03           33.86%    passed     98.35%\n",
      " 453  Pseudomonas_aeruginosa_002085605_00           33.81%    passed     97.51%\n",
      " 454  Pseudomonas_aeruginosa_002085605_01           33.84%    passed     98.02%\n",
      " 455  Pseudomonas_aeruginosa_002085605_02           33.84%    passed     97.53%\n",
      " 456  Pseudomonas_aeruginosa_002085605_03           33.83%    passed     97.67%\n",
      " 457  Pseudomonas_koreensis_003049825_00            30.35%    passed     95.29%\n",
      " 458  Pseudomonas_koreensis_003049825_01            33.15%    passed     98.42%\n",
      " 459  Pseudomonas_koreensis_003049825_02            32.93%    passed     96.35%\n",
      " 460  Pseudomonas_koreensis_003049825_03            35.96%    passed     83.39%\n",
      " 461  Pseudomonas_koreensis_003049825_04            33.13%    passed     98.34%\n",
      " 462  Pseudomonas_koreensis_003049825_05            32.85%    passed     97.15%\n",
      " 463  Pseudomonas_synxantha_003851525_00            30.07%    passed     76.06%\n",
      " 464  Pseudomonas_synxantha_003851525_01            32.95%    passed     83.52%\n",
      " 465  Pseudomonas_synxantha_003851525_02            32.95%    passed     81.32%\n",
      " 466  Pseudomonas_synxantha_003851525_03            32.95%    passed     83.10%\n",
      " 467  Pseudomonas_synxantha_003851525_04            32.07%    passed     90.62%\n",
      " 468  Pseudomonas_synxantha_003851525_05            33.07%    passed     92.18%\n",
      " 469  Pseudomonas_mendocina_000204295_00            33.31%    passed     99.57%\n",
      " 470  Pseudomonas_mendocina_000204295_01            33.31%    passed     99.69%\n",
      " 471  Pseudomonas_mendocina_000204295_02            33.31%    passed     99.80%\n",
      " 472  Pseudomonas_mendocina_000204295_03            33.31%    passed     99.69%\n",
      " 473  Pseudomonas_orientalis_003851645_00           30.19%    passed     88.78%\n",
      " 474  Pseudomonas_orientalis_003851645_01           33.25%    passed     96.30%\n",
      " 475  Pseudomonas_orientalis_003851645_02           32.67%    passed     89.24%\n",
      " 476  Pseudomonas_orientalis_003851645_03           32.68%    passed     92.48%\n",
      " 477  Pseudomonas_orientalis_003851645_04           32.93%    passed     85.22%\n",
      " 478  Pseudomonas_orientalis_003851645_05           32.94%    passed     93.24%\n",
      " 479  Pseudomonas_aeruginosa_003319235_00           33.83%    passed     97.94%\n",
      " 480  Pseudomonas_aeruginosa_003319235_01           33.84%    passed     98.06%\n",
      " 481  Pseudomonas_aeruginosa_003319235_02           33.81%    passed     99.02%\n",
      " 482  Pseudomonas_aeruginosa_003319235_03           33.83%    passed     97.35%\n",
      " 483  Pseudomonas_aeruginosa_900149285_00           33.83%    passed     97.67%\n",
      " 484  Pseudomonas_aeruginosa_900149285_01           33.84%    passed     96.95%\n",
      " 485  Pseudomonas_aeruginosa_900149285_02           33.83%    passed     97.67%\n",
      " 486  Pseudomonas_aeruginosa_900149285_03           33.83%    passed     97.67%\n",
      " 487  Pseudomonas_aeruginosa_900618245_00           33.81%    passed     99.15%\n",
      " 488  Pseudomonas_aeruginosa_900618245_01           33.81%    passed     99.15%\n",
      " 489  Pseudomonas_aeruginosa_900618245_03           59.48%    passed     89.14%\n",
      " 490  Pseudomonas_aeruginosa_900618245_04           33.81%    passed     99.15%\n",
      " 491  Pseudomonas_aeruginosa_000829275_00           33.84%    passed     99.04%\n",
      " 492  Pseudomonas_aeruginosa_000829275_01           33.84%    passed     99.04%\n",
      " 493  Pseudomonas_aeruginosa_000829275_02           33.84%    passed     98.51%\n",
      " 494  Pseudomonas_aeruginosa_000829275_03           33.84%    passed     99.04%\n",
      " 495  Pseudomonas_syringae_002966555_00             32.73%    passed     97.71%\n",
      " 496  Pseudomonas_syringae_002966555_01             32.73%    passed     97.29%\n",
      " 497  Pseudomonas_syringae_002966555_02             32.72%    passed     97.50%\n",
      " 498  Pseudomonas_syringae_002966555_03             32.73%    passed     97.71%\n",
      " 499  Pseudomonas_syringae_002966555_04             29.87%    passed     99.01%\n",
      " 500  Pseudomonas_aeruginosa_003952325_00           33.77%    passed     98.64%\n",
      " 501  Pseudomonas_aeruginosa_003952325_01           33.74%    passed     98.20%\n",
      " 502  Pseudomonas_aeruginosa_003952325_02           33.80%    passed     97.90%\n",
      " 503  Pseudomonas_aeruginosa_003952325_03           33.83%    passed     98.41%\n",
      " 504  Pseudomonas_putida_003228315_00               33.32%    passed     99.14%\n",
      " 505  Pseudomonas_putida_003228315_01               33.32%    passed     98.47%\n",
      " 506  Pseudomonas_putida_003228315_02               33.32%    passed     98.47%\n",
      " 507  Pseudomonas_putida_003228315_03               33.32%    passed     99.14%\n",
      " 508  Pseudomonas_putida_003228315_04               30.48%    passed     96.66%\n",
      " 509  Pseudomonas_putida_003228315_05               33.32%    passed     98.47%\n",
      " 510  Pseudomonas_aeruginosa_900069025_02           33.81%    passed     98.49%\n",
      " 511  Pseudomonas_aeruginosa_900069025_03           33.81%    passed     98.49%\n",
      " 512  Pseudomonas_aeruginosa_900069025_04           33.81%    passed     98.49%\n",
      " 513  Pseudomonas_putida_000410575_00               33.81%    passed     97.41%\n",
      " 514  Pseudomonas_putida_000410575_01               36.35%    passed     99.87%\n",
      " 515  Pseudomonas_putida_000410575_02               33.40%    passed     99.32%\n",
      " 516  Pseudomonas_putida_000410575_03               33.41%    passed     99.54%\n",
      " 517  Pseudomonas_putida_000410575_04               59.55%    passed     91.11%\n",
      " 518  Pseudomonas_putida_000410575_05               32.53%    passed     96.57%\n",
      " 519  Pseudomonas_stutzeri_000661915_00             32.99%    passed     99.76%\n",
      " 520  Pseudomonas_stutzeri_000661915_01             32.99%    passed     99.76%\n",
      " 521  Pseudomonas_stutzeri_000661915_02             32.99%    passed     99.76%\n",
      " 522  Pseudomonas_stutzeri_000661915_03             32.99%    passed     99.76%\n",
      " 523  Pseudomonas_aeruginosa_000271365_00           33.83%    passed     98.77%\n",
      " 524  Pseudomonas_aeruginosa_000271365_01           33.83%    passed     98.40%\n",
      " 525  Pseudomonas_aeruginosa_000271365_02           33.83%    passed     98.40%\n",
      " 526  Pseudomonas_aeruginosa_000271365_03           33.83%    passed     98.40%\n",
      " 527  Pseudomonas_aeruginosa_002215345_00           33.83%    passed     98.25%\n",
      " 528  Pseudomonas_aeruginosa_002215345_01           33.83%    passed     98.15%\n",
      " 529  Pseudomonas_aeruginosa_002215345_02           33.83%    passed     97.67%\n",
      " 530  Pseudomonas_aeruginosa_002215345_03           33.83%    passed     98.25%\n",
      " 531  Pseudomonas_aeruginosa_002968515_00           33.83%    passed     98.25%\n",
      " 532  Pseudomonas_aeruginosa_002968515_01           33.83%    passed     97.67%\n",
      " 533  Pseudomonas_aeruginosa_002968515_02           33.83%    passed     97.80%\n",
      " 534  Pseudomonas_aeruginosa_002968515_03           33.83%    passed     98.25%\n",
      " 535  Pseudomonas_aeruginosa_001516225_00           33.84%    passed     97.82%\n",
      " 536  Pseudomonas_aeruginosa_001516225_01           33.81%    passed     98.22%\n",
      " 537  Pseudomonas_aeruginosa_001516225_02           33.83%    passed     98.75%\n",
      " 538  Pseudomonas_aeruginosa_001516225_03           33.83%    passed     98.23%\n",
      " 539  Pseudomonas_aeruginosa_003028335_00           33.84%    passed     99.04%\n",
      " 540  Pseudomonas_aeruginosa_003028335_01           33.84%    passed     99.04%\n",
      " 541  Pseudomonas_aeruginosa_003028335_02           33.84%    passed     98.81%\n",
      " 542  Pseudomonas_aeruginosa_003028335_03           33.85%    passed     99.11%\n",
      " 543  Pseudomonas_aeruginosa_000508765_00           33.83%    passed     98.40%\n",
      " 544  Pseudomonas_aeruginosa_000508765_01           33.80%    passed     98.07%\n",
      " 545  Pseudomonas_aeruginosa_000508765_02           34.38%    passed     95.92%\n",
      " 546  Pseudomonas_aeruginosa_000508765_03           33.84%    passed     98.88%\n",
      " 547  Pseudomonas_putida_900636645_01               36.25%    passed     99.23%\n",
      " 548  Pseudomonas_putida_900636645_02               33.36%    passed     99.01%\n",
      " 549  Pseudomonas_putida_900636645_03               36.10%    passed     98.90%\n",
      " 550  Pseudomonas_putida_900636645_04               33.36%    passed     98.89%\n",
      " 551  Pseudomonas_putida_900636645_05               32.53%    passed     96.57%\n",
      " 552  Pseudomonas_orientalis_003852045_00           30.05%    passed     78.74%\n",
      " 553  Pseudomonas_orientalis_003852045_01           32.79%    passed     84.21%\n",
      " 554  Pseudomonas_orientalis_003852045_02           33.08%    passed     95.94%\n",
      " 555  Pseudomonas_orientalis_003852045_03           32.94%    passed     92.26%\n",
      " 556  Pseudomonas_orientalis_003852045_04           32.95%    passed     92.29%\n",
      " 557  Pseudomonas_orientalis_003852045_05           32.71%    passed     88.62%\n",
      " 558  Pseudomonas_aeruginosa_001548335_00           33.83%    passed     96.98%\n",
      " 559  Pseudomonas_aeruginosa_001548335_01           33.83%    passed     97.61%\n",
      " 560  Pseudomonas_aeruginosa_001548335_02           33.83%    passed     98.40%\n",
      " 561  Pseudomonas_aeruginosa_001548335_03           33.83%    passed     98.40%\n",
      " 562  Pseudomonas_aeruginosa_002237405_00           33.84%    passed     98.29%\n",
      " 563  Pseudomonas_aeruginosa_002237405_01           33.84%    passed     97.64%\n",
      " 564  Pseudomonas_aeruginosa_002237405_02           33.84%    passed     97.53%\n",
      " 565  Pseudomonas_aeruginosa_002237405_03           33.85%    passed     97.69%\n",
      " 566  Pseudomonas_stutzeri_000267545_00             33.21%    passed     99.67%\n",
      " 567  Pseudomonas_stutzeri_000267545_01             33.21%    passed     99.49%\n",
      " 568  Pseudomonas_stutzeri_000267545_02             33.20%    passed     99.47%\n",
      " 569  Pseudomonas_stutzeri_000267545_03             33.21%    passed     99.49%\n",
      " 570  Pseudomonas_aeruginosa_001721805_00           33.83%    passed     98.00%\n",
      " 571  Pseudomonas_aeruginosa_001721805_01           33.81%    passed     97.59%\n",
      " 572  Pseudomonas_aeruginosa_001721805_02           33.81%    passed     98.30%\n",
      " 573  Pseudomonas_aeruginosa_001721805_03           33.84%    passed     97.82%\n",
      " 574  Pseudomonas_aeruginosa_003571805_00           33.84%    passed     98.29%\n",
      " 575  Pseudomonas_aeruginosa_003571805_01           33.83%    passed     97.67%\n",
      " 576  Pseudomonas_aeruginosa_003571805_02           33.85%    passed     98.50%\n",
      " 577  Pseudomonas_aeruginosa_003571805_03           33.84%    passed     98.38%\n",
      " 578  Pseudomonas_syringae_002905815_00             32.77%    passed     98.80%\n",
      " 579  Pseudomonas_syringae_002905815_01             32.77%    passed     98.76%\n",
      " 580  Pseudomonas_syringae_002905815_02             32.57%    passed     99.99%\n",
      " 581  Pseudomonas_syringae_002905815_03             32.77%    passed     98.97%\n",
      " 582  Pseudomonas_syringae_002905815_04             32.77%    passed     99.03%\n",
      " 583  Pseudomonas_aeruginosa_003991465_00           33.79%    passed     97.99%\n",
      " 584  Pseudomonas_aeruginosa_003991465_01           33.80%    passed     97.80%\n",
      " 585  Pseudomonas_aeruginosa_003991465_02           33.80%    passed     97.38%\n",
      " 586  Pseudomonas_aeruginosa_003991465_03           33.80%    passed     98.11%\n",
      " 587  Pseudomonas_aeruginosa_000981825_00           33.83%    passed     97.35%\n",
      " 588  Pseudomonas_aeruginosa_000981825_01           33.83%    passed     97.35%\n",
      " 589  Pseudomonas_aeruginosa_000981825_02           33.83%    passed     97.35%\n",
      " 590  Pseudomonas_aeruginosa_000981825_03           33.83%    passed     97.35%\n",
      " 591  Pseudomonas_synxantha_003851465_00            30.07%    passed     69.54%\n",
      " 592  Pseudomonas_synxantha_003851465_01            33.09%    passed     93.73%\n",
      " 593  Pseudomonas_synxantha_003851465_02            32.95%    passed     84.36%\n",
      " 594  Pseudomonas_synxantha_003851465_03            33.11%    passed     94.24%\n",
      " 595  Pseudomonas_synxantha_003851465_04            32.97%    passed     82.41%\n",
      " 596  Pseudomonas_synxantha_003851465_05            32.09%    passed     92.12%\n",
      " 597  Pseudomonas_aeruginosa_003369755_00           33.81%    passed     97.51%\n",
      " 598  Pseudomonas_aeruginosa_003369755_01           33.81%    passed     97.51%\n",
      " 599  Pseudomonas_aeruginosa_003369755_02           33.83%    passed     97.93%\n",
      " 600  Pseudomonas_aeruginosa_003369755_03           33.81%    passed     97.51%\n",
      " 601  Pseudomonas_aeruginosa_002812825_00           33.83%    passed     98.55%\n",
      " 602  Pseudomonas_aeruginosa_002812825_01           33.83%    passed     97.94%\n",
      " 603  Pseudomonas_aeruginosa_002812825_02           33.83%    passed     98.40%\n",
      " 604  Pseudomonas_aeruginosa_002812825_03           33.83%    passed     98.77%\n",
      " 605  Pseudomonas_fluorescens_900636635_00          33.83%    passed     98.96%\n",
      " 606  Pseudomonas_fluorescens_900636635_01          33.83%    passed     98.40%\n",
      " 607  Pseudomonas_fluorescens_900636635_02          33.83%    passed     98.41%\n",
      " 608  Pseudomonas_fluorescens_900636635_03          33.83%    passed     98.59%\n",
      " 609  Pseudomonas_putida_000219705_00               33.06%    passed     99.98%\n",
      " 610  Pseudomonas_putida_000219705_01               35.92%    passed     98.79%\n",
      " 611  Pseudomonas_putida_000219705_02               35.92%    passed     99.01%\n",
      " 612  Pseudomonas_putida_000219705_03               35.92%    passed     99.59%\n",
      " 613  Pseudomonas_putida_000219705_04               35.88%    passed     98.44%\n",
      " 614  Pseudomonas_putida_000219705_05               35.92%    passed     99.25%\n",
      " 615  Pseudomonas_aeruginosa_001548135_00           33.81%    passed     98.65%\n",
      " 616  Pseudomonas_aeruginosa_001548135_01           33.83%    passed     97.18%\n",
      " 617  Pseudomonas_aeruginosa_001548135_02           33.84%    passed     98.51%\n",
      " 618  Pseudomonas_aeruginosa_001548135_03           33.81%    passed     98.65%\n",
      " 619  Pseudomonas_monteilii_000510285_00            35.93%    passed     97.93%\n",
      " 620  Pseudomonas_monteilii_000510285_01            35.93%    passed     98.42%\n",
      " 621  Pseudomonas_monteilii_000510285_02            35.93%    passed     98.85%\n",
      " 622  Pseudomonas_monteilii_000510285_03            35.93%    passed     98.14%\n",
      " 623  Pseudomonas_monteilii_000510285_04            35.93%    passed     98.85%\n",
      " 624  Pseudomonas_monteilii_000510285_05            33.08%    passed     99.41%\n",
      " 625  Pseudomonas_aeruginosa_001618925_00           33.83%    passed     97.67%\n",
      " 626  Pseudomonas_aeruginosa_001618925_01           33.81%    passed     98.52%\n",
      " 627  Pseudomonas_aeruginosa_001618925_02           33.81%    passed     98.27%\n",
      " 628  Pseudomonas_aeruginosa_001618925_03           33.83%    passed     97.67%\n",
      " 629  Pseudomonas_fluorescens_000293885_01          33.29%    passed     98.52%\n",
      " 630  Pseudomonas_fluorescens_000293885_02          30.26%    passed     89.04%\n",
      " 631  Pseudomonas_fluorescens_000293885_03          33.62%    passed     97.12%\n",
      " 632  Pseudomonas_fluorescens_000293885_04          33.09%    passed     97.72%\n",
      " 633  Pseudomonas_fluorescens_000293885_05          33.29%    passed     98.53%\n",
      " 634  Pseudomonas_fluorescens_000293885_06          59.27%    passed     74.21%\n",
      " 635  Pseudomonas_aeruginosa_900095805_00           33.83%    passed     98.40%\n",
      " 636  Pseudomonas_aeruginosa_900095805_01           33.81%    passed     98.49%\n",
      " 637  Pseudomonas_aeruginosa_900095805_02           33.83%    passed     97.94%\n",
      " 638  Pseudomonas_aeruginosa_900095805_03           33.81%    passed     98.49%\n",
      " 639  Pseudomonas_aeruginosa_001722005_00           33.84%    passed     97.64%\n",
      " 640  Pseudomonas_aeruginosa_001722005_01           33.84%    passed     98.29%\n",
      " 641  Pseudomonas_aeruginosa_001722005_02           33.84%    passed     97.64%\n",
      " 642  Pseudomonas_aeruginosa_001722005_03           33.84%    passed     98.38%\n",
      " 643  Pseudomonas_mendocina_003008615_00            33.27%    passed     99.04%\n",
      " 644  Pseudomonas_mendocina_003008615_01            33.27%    passed     99.04%\n",
      " 645  Pseudomonas_mendocina_003008615_02            33.27%    passed     99.04%\n",
      " 646  Pseudomonas_mendocina_003008615_03            33.27%    passed     99.04%\n",
      " 647  Pseudomonas_syringae_000344355_00             32.79%    passed     97.23%\n",
      " 648  Pseudomonas_syringae_000344355_01             32.79%    passed     97.23%\n",
      " 649  Pseudomonas_syringae_000344355_02             32.79%    passed     97.63%\n",
      " 650  Pseudomonas_syringae_000344355_03             32.79%    passed     97.23%\n",
      " 651  Pseudomonas_syringae_000344355_04             29.92%    passed     98.24%\n",
      " 652  Pseudomonas_putida_000019445_00               33.59%    passed     96.50%\n",
      " 653  Pseudomonas_putida_000019445_01               33.94%    passed     98.79%\n",
      " 654  Pseudomonas_putida_000019445_02               33.94%    passed     98.79%\n",
      " 655  Pseudomonas_putida_000019445_03               33.59%    passed     95.99%\n",
      " 656  Pseudomonas_putida_000019445_04               35.92%    passed     94.99%\n",
      " 657  Pseudomonas_chlororaphis_003850485_00         30.82%    passed     95.61%\n",
      " 658  Pseudomonas_chlororaphis_003850485_01         33.71%    passed     96.71%\n",
      " 659  Pseudomonas_chlororaphis_003850485_02         33.57%    passed     96.73%\n",
      " 660  Pseudomonas_chlororaphis_003850485_03         33.71%    passed     97.31%\n",
      " 661  Pseudomonas_chlororaphis_003850485_04         33.71%    passed     96.71%\n",
      " 662  Pseudomonas_aeruginosa_900636755_00           33.83%    passed     98.15%\n",
      " 663  Pseudomonas_aeruginosa_900636755_01           33.83%    passed     98.40%\n",
      " 664  Pseudomonas_aeruginosa_900636755_02           33.81%    passed     98.71%\n",
      " 665  Pseudomonas_aeruginosa_900636755_03           33.81%    passed     98.27%\n",
      " 666  Pseudomonas_fluorescens_001747385_00          33.17%    passed     87.56%\n",
      " 667  Pseudomonas_fluorescens_001747385_01          33.13%    passed     86.74%\n",
      " 668  Pseudomonas_fluorescens_001747385_02          32.95%    passed     77.77%\n",
      " 669  Pseudomonas_fluorescens_001747385_03          33.11%    passed     84.73%\n",
      " 670  Pseudomonas_fluorescens_001747385_04          30.22%    passed     76.08%\n",
      " 671  Pseudomonas_fluorescens_001747385_05          33.12%    passed     87.12%\n",
      " 672  Pseudomonas_putida_000019125_01               36.35%    passed     99.87%\n",
      " 673  Pseudomonas_putida_000019125_02               33.40%    passed     99.32%\n",
      " 674  Pseudomonas_putida_000019125_03               33.40%    passed     99.53%\n",
      " 675  Pseudomonas_putida_000019125_04               36.35%    passed     99.92%\n",
      " 676  Pseudomonas_putida_000019125_05               36.35%    passed     99.85%\n",
      " 677  Pseudomonas_aeruginosa_002104595_00           33.83%    passed     98.59%\n",
      " 678  Pseudomonas_aeruginosa_002104595_01           33.84%    passed     98.06%\n",
      " 679  Pseudomonas_aeruginosa_002104595_02           33.83%    passed     98.15%\n",
      " 680  Pseudomonas_aeruginosa_002104595_03           33.84%    passed     98.25%\n",
      " 681  Pseudomonas_aeruginosa_003073695_00           33.89%    passed     98.36%\n",
      " 682  Pseudomonas_aeruginosa_003073695_01           33.90%    passed     99.19%\n",
      " 683  Pseudomonas_aeruginosa_003073695_02           33.84%    passed     97.58%\n",
      " 684  Pseudomonas_aeruginosa_003073695_03           33.90%    passed     98.00%\n",
      " 685  Pseudomonas_putida_000691565_01               35.93%    passed     98.42%\n",
      " 686  Pseudomonas_putida_000691565_02               35.96%    passed     99.68%\n",
      " 687  Pseudomonas_putida_000691565_03               35.93%    passed     97.64%\n",
      " 688  Pseudomonas_putida_000691565_04               35.88%    passed     98.44%\n",
      " 689  Pseudomonas_putida_000691565_05               33.50%    passed     99.15%\n",
      " 690  Pseudomonas_aeruginosa_003073615_00           33.83%    passed     98.15%\n",
      " 691  Pseudomonas_aeruginosa_003073615_01           33.83%    passed     98.25%\n",
      " 692  Pseudomonas_aeruginosa_003073615_02           33.83%    passed     98.25%\n",
      " 693  Pseudomonas_aeruginosa_003073615_03           33.83%    passed     97.67%\n",
      " 694  Pseudomonas_aeruginosa_002812905_00           33.84%    passed     99.04%\n",
      " 695  Pseudomonas_aeruginosa_002812905_01           33.84%    passed     99.04%\n",
      " 696  Pseudomonas_aeruginosa_002812905_02           33.85%    passed     98.61%\n",
      " 697  Pseudomonas_aeruginosa_002812905_03           33.84%    passed     99.04%\n",
      " 698  Pseudomonas_syringae_002024305_00             32.79%    passed     97.23%\n",
      " 699  Pseudomonas_syringae_002024305_01             32.79%    passed     97.23%\n",
      " 700  Pseudomonas_syringae_002024305_02             32.79%    passed     97.23%\n",
      " 701  Pseudomonas_syringae_002024305_03             32.79%    passed     97.63%\n",
      " 702  Pseudomonas_syringae_002024305_04             29.92%    passed     98.24%\n",
      " 703  Pseudomonas_aeruginosa_900618305_00           35.33%    passed     94.68%\n",
      " 704  Pseudomonas_aeruginosa_900618305_01           33.88%    passed     98.20%\n",
      " 705  Pseudomonas_aeruginosa_900618305_02           33.83%    passed     97.72%\n",
      " 706  Pseudomonas_aeruginosa_900618305_03           33.88%    passed     98.75%\n",
      " 707  Pseudomonas_aeruginosa_002223805_00           33.84%    passed     99.04%\n",
      " 708  Pseudomonas_aeruginosa_002223805_01           33.81%    passed     98.03%\n",
      " 709  Pseudomonas_aeruginosa_002223805_02           33.84%    passed     98.06%\n",
      " 710  Pseudomonas_aeruginosa_002223805_03           33.81%    passed     98.86%\n",
      " 711  Pseudomonas_aeruginosa_001792835_00           33.81%    passed     98.27%\n",
      " 712  Pseudomonas_aeruginosa_001792835_01           33.81%    passed     98.27%\n",
      " 713  Pseudomonas_aeruginosa_001792835_02           33.83%    passed     97.67%\n",
      " 714  Pseudomonas_aeruginosa_001792835_03           33.83%    passed     98.15%\n",
      " 715  Pseudomonas_synxantha_003851495_00            30.18%    passed     77.86%\n",
      " 716  Pseudomonas_synxantha_003851495_01            33.45%    passed     85.79%\n",
      " 717  Pseudomonas_synxantha_003851495_02            32.97%    passed     80.19%\n",
      " 718  Pseudomonas_synxantha_003851495_03            32.20%    passed     88.46%\n",
      " 719  Pseudomonas_synxantha_003851495_04            32.97%    passed     80.19%\n",
      " 720  Pseudomonas_synxantha_003851495_05            32.97%    passed     78.85%\n",
      " 721  Pseudomonas_putida_000007565_01               36.42%    passed     99.57%\n",
      " 722  Pseudomonas_putida_000007565_02               33.53%    passed     98.40%\n",
      " 723  Pseudomonas_putida_000007565_03               36.26%    passed     99.19%\n",
      " 724  Pseudomonas_putida_000007565_04               33.53%    passed     98.36%\n",
      " 725  Pseudomonas_putida_000007565_05               32.70%    passed     95.34%\n",
      " 726  Pseudomonas_aeruginosa_002997005_00           33.83%    passed     98.40%\n",
      " 727  Pseudomonas_aeruginosa_002997005_01           33.84%    passed     97.26%\n",
      " 728  Pseudomonas_aeruginosa_002997005_02           33.84%    passed     97.26%\n",
      " 729  Pseudomonas_aeruginosa_002997005_03           33.84%    passed     98.25%\n",
      " 730  Pseudomonas_syringae_900235905_00             32.79%    passed     97.99%\n",
      " 731  Pseudomonas_syringae_900235905_01             32.79%    passed     97.99%\n",
      " 732  Pseudomonas_syringae_900235905_02             32.79%    passed     97.99%\n",
      " 733  Pseudomonas_syringae_900235905_03             32.79%    passed     98.38%\n",
      " 734  Pseudomonas_syringae_900235905_04             29.92%    passed     97.75%\n",
      " 735  Pseudomonas_parafulva_003410295_00            30.23%    passed     91.66%\n",
      " 736  Pseudomonas_parafulva_003410295_01            33.12%    passed     92.86%\n",
      " 737  Pseudomonas_parafulva_003410295_02            33.12%    passed     91.33%\n",
      " 738  Pseudomonas_parafulva_003410295_03            33.12%    passed     91.33%\n",
      " 739  Pseudomonas_parafulva_003410295_04            33.13%    passed     92.65%\n",
      " 740  Pseudomonas_parafulva_003410295_05            33.12%    passed     92.14%\n",
      " 741  Pseudomonas_aeruginosa_001457615_00           33.85%    passed     98.61%\n",
      " 742  Pseudomonas_aeruginosa_001457615_01           33.79%    passed     98.72%\n",
      " 743  Pseudomonas_aeruginosa_001457615_02           33.83%    passed     98.77%\n",
      " 744  Pseudomonas_aeruginosa_001457615_03           33.84%    passed     98.68%\n",
      " 745  Pseudomonas_aeruginosa_001632245_00           33.85%    passed     97.40%\n",
      " 746  Pseudomonas_aeruginosa_001632245_01           33.84%    passed     96.95%\n",
      " 747  Pseudomonas_aeruginosa_001632245_02           33.83%    passed     98.65%\n",
      " 748  Pseudomonas_aeruginosa_001632245_03           33.83%    passed     98.15%\n",
      " 749  Pseudomonas_aeruginosa_002812865_00           33.83%    passed     97.93%\n",
      " 750  Pseudomonas_aeruginosa_002812865_01           33.83%    passed     97.93%\n",
      " 751  Pseudomonas_aeruginosa_002812865_02           33.83%    passed     97.55%\n",
      " 752  Pseudomonas_aeruginosa_002812865_03           33.83%    passed     97.85%\n",
      " 753  Pseudomonas_fluorescens_000934565_00          30.08%    passed     51.32%\n",
      " 754  Pseudomonas_fluorescens_000934565_01          33.08%    passed     71.60%\n",
      " 755  Pseudomonas_fluorescens_000934565_02          32.97%    passed     64.73%\n",
      " 756  Pseudomonas_fluorescens_000934565_03          35.66%    passed     57.21%\n",
      " 757  Pseudomonas_fluorescens_000934565_04          32.97%    passed     69.05%\n",
      " 758  Pseudomonas_fluorescens_000934565_05          32.97%    passed     63.60%\n",
      " 759  Pseudomonas_syringae_000344335_00             32.79%    passed     97.66%\n",
      " 760  Pseudomonas_syringae_000344335_01             32.79%    passed     97.19%\n",
      " 761  Pseudomonas_syringae_000344335_02             32.79%    passed     97.60%\n",
      " 762  Pseudomonas_syringae_000344335_03             32.79%    passed     97.66%\n",
      " 763  Pseudomonas_syringae_000344335_04             29.92%    passed     98.82%\n",
      " 764  Pseudomonas_aeruginosa_002968695_00           33.83%    passed     98.15%\n",
      " 765  Pseudomonas_aeruginosa_002968695_01           33.83%    passed     97.67%\n",
      " 766  Pseudomonas_aeruginosa_002968695_02           33.84%    passed     98.38%\n",
      " 767  Pseudomonas_aeruginosa_002968695_03           33.83%    passed     98.25%\n",
      " 768  Pseudomonas_chlororaphis_003851835_00         30.67%    passed     92.49%\n",
      " 769  Pseudomonas_chlororaphis_003851835_01         33.57%    passed     97.04%\n",
      " 770  Pseudomonas_chlororaphis_003851835_02         33.57%    passed     97.04%\n",
      " 771  Pseudomonas_chlororaphis_003851835_03         33.57%    passed     97.04%\n",
      " 772  Pseudomonas_chlororaphis_003851835_04         33.57%    passed     97.04%\n",
      " 773  Pseudomonas_aeruginosa_002968755_00           33.85%    passed     97.75%\n",
      " 774  Pseudomonas_aeruginosa_002968755_01           33.85%    passed     97.77%\n",
      " 775  Pseudomonas_aeruginosa_002968755_02           33.84%    passed     98.03%\n",
      " 776  Pseudomonas_aeruginosa_002968755_03           33.85%    passed     98.79%\n",
      " 777  Pseudomonas_aeruginosa_001516185_00           33.83%    passed     98.00%\n",
      " 778  Pseudomonas_aeruginosa_001516185_01           33.84%    passed     97.10%\n",
      " 779  Pseudomonas_aeruginosa_001516185_02           33.84%    passed     97.36%\n",
      " 780  Pseudomonas_aeruginosa_001516185_03           33.84%    passed     97.95%\n",
      " 781  Pseudomonas_entomophila_000026105_00          30.87%    passed     97.57%\n",
      " 782  Pseudomonas_entomophila_000026105_01          33.79%    passed     99.96%\n",
      " 783  Pseudomonas_entomophila_000026105_02          33.77%    passed     99.83%\n",
      " 784  Pseudomonas_entomophila_000026105_03          33.79%    passed     99.89%\n",
      " 785  Pseudomonas_entomophila_000026105_04          33.79%    passed     99.84%\n",
      " 786  Pseudomonas_entomophila_000026105_05          33.79%    passed     99.90%\n",
      " 787  Pseudomonas_entomophila_000026105_06          33.41%    passed     98.72%\n",
      " 788  Pseudomonas_chlororaphis_900636995_00         33.72%    passed     97.25%\n",
      " 789  Pseudomonas_chlororaphis_900636995_01         33.72%    passed     96.63%\n",
      " 790  Pseudomonas_chlororaphis_900636995_02         33.72%    passed     96.63%\n",
      " 791  Pseudomonas_chlororaphis_900636995_03         30.82%    passed     93.87%\n",
      " 792  Pseudomonas_chlororaphis_900636995_04         33.72%    passed     96.63%\n",
      " 793  Pseudomonas_fluorescens_001708445_00          33.08%    passed     94.76%\n",
      " 794  Pseudomonas_syringae_002763655_00             30.04%    passed     96.79%\n",
      " 795  Pseudomonas_syringae_002763655_01             32.80%    passed     97.00%\n",
      " 796  Pseudomonas_syringae_002763655_02             32.97%    passed     99.12%\n",
      " 797  Pseudomonas_syringae_002763655_03             32.85%    passed     97.76%\n",
      " 798  Pseudomonas_syringae_002763655_04             32.90%    passed     98.46%\n",
      " 799  Pseudomonas_brassicacearum_001746815_00       32.75%    passed     99.81%\n",
      " 800  Pseudomonas_brassicacearum_001746815_01       30.10%    passed     99.27%\n",
      " 801  Pseudomonas_brassicacearum_001746815_02       32.44%    passed     99.99%\n",
      " 802  Pseudomonas_brassicacearum_001746815_03       32.75%    passed     99.94%\n",
      " 803  Pseudomonas_brassicacearum_001746815_04       32.75%    passed     99.89%\n",
      " 804  Pseudomonas_chlororaphis_003851165_00         30.82%    passed     94.73%\n",
      " 805  Pseudomonas_chlororaphis_003851165_01         33.72%    passed     96.65%\n",
      " 806  Pseudomonas_chlororaphis_003851165_02         33.72%    passed     97.33%\n",
      " 807  Pseudomonas_chlororaphis_003851165_03         33.72%    passed     96.65%\n",
      " 808  Pseudomonas_chlororaphis_003851165_04         33.72%    passed     97.33%\n",
      " 809  Pseudomonas_putida_000226035_00               56.76%    passed     87.08%\n",
      " 810  Pseudomonas_putida_000226035_01               33.97%    passed     99.94%\n",
      " 811  Pseudomonas_putida_000226035_02               10.79%    passed     97.89%\n",
      " 812  Pseudomonas_putida_000226035_03               36.25%    passed     99.03%\n",
      " 813  Pseudomonas_putida_000226035_04               59.63%    passed     79.07%\n",
      " 814  Pseudomonas_putida_000226035_06               33.95%    passed     99.70%\n",
      " 815  Pseudomonas_putida_000226035_07               36.98%    passed     99.81%\n",
      " 816  Pseudomonas_aeruginosa_003060845_00           33.74%    passed     91.59%\n",
      " 817  Pseudomonas_aeruginosa_003060845_01           33.74%    passed     97.92%\n",
      " 818  Pseudomonas_aeruginosa_003060845_02           33.75%    passed     98.43%\n",
      " 819  Pseudomonas_aeruginosa_003060845_03           33.76%    passed     98.14%\n",
      " 820  Pseudomonas_aeruginosa_004010895_00           33.83%    passed     98.50%\n",
      " 821  Pseudomonas_aeruginosa_004010895_01           33.84%    passed     98.21%\n",
      " 822  Pseudomonas_aeruginosa_004010895_02           33.85%    passed     98.45%\n",
      " 823  Pseudomonas_aeruginosa_004010895_03           33.84%    passed     98.35%\n",
      " 824  Pseudomonas_aeruginosa_003369775_00           33.84%    passed     98.29%\n",
      " 825  Pseudomonas_aeruginosa_003369775_01           33.84%    passed     98.29%\n",
      " 826  Pseudomonas_aeruginosa_003369775_02           33.84%    passed     98.29%\n",
      " 827  Pseudomonas_aeruginosa_003369775_03           33.84%    passed     97.95%\n",
      " 828  Pseudomonas_aeruginosa_001447845_00           33.83%    passed     96.77%\n",
      " 829  Pseudomonas_aeruginosa_001447845_01           33.83%    passed     96.77%\n",
      " 830  Pseudomonas_aeruginosa_001447845_02           33.84%    passed     96.95%\n",
      " 831  Pseudomonas_aeruginosa_001447845_03           33.84%    passed     97.18%\n",
      " 832  Pseudomonas_aeruginosa_001516345_00           33.83%    passed     97.61%\n",
      " 833  Pseudomonas_aeruginosa_001516345_01           33.84%    passed     97.82%\n",
      " 834  Pseudomonas_aeruginosa_001516345_02           33.84%    passed     98.29%\n",
      " 835  Pseudomonas_aeruginosa_001516345_03           33.83%    passed     98.40%\n",
      " 836  Pseudomonas_koreensis_001654515_00            30.39%    passed     93.85%\n",
      " 837  Pseudomonas_koreensis_001654515_01            33.25%    passed     97.02%\n",
      " 838  Pseudomonas_koreensis_001654515_02            33.25%    passed     98.07%\n",
      " 839  Pseudomonas_koreensis_001654515_03            32.98%    passed     94.71%\n",
      " 840  Pseudomonas_koreensis_001654515_04            33.25%    passed     98.20%\n",
      " 841  Pseudomonas_koreensis_001654515_05            33.26%    passed     98.70%\n",
      " 842  Pseudomonas_synxantha_003852025_00            29.18%    passed     89.48%\n",
      " 843  Pseudomonas_synxantha_003852025_01            32.84%    passed     88.41%\n",
      " 844  Pseudomonas_synxantha_003852025_02            32.07%    passed     88.54%\n",
      " 845  Pseudomonas_synxantha_003852025_03            33.08%    passed     91.39%\n",
      " 846  Pseudomonas_synxantha_003852025_04            33.08%    passed     94.04%\n",
      " 847  Pseudomonas_synxantha_003852025_05            32.84%    passed     88.41%\n",
      " 848  Pseudomonas_monteilii_003671975_01            36.38%    passed     94.05%\n",
      " 849  Pseudomonas_monteilii_003671975_02            36.38%    passed     94.05%\n",
      " 850  Pseudomonas_monteilii_003671975_03            36.38%    passed     94.05%\n",
      " 851  Pseudomonas_monteilii_003671975_04            33.48%    passed     96.79%\n",
      " 852  Pseudomonas_monteilii_003671975_05            33.48%    passed     96.79%\n",
      " 853  Pseudomonas_chlororaphis_003851305_00         30.67%    passed     92.49%\n",
      " 854  Pseudomonas_chlororaphis_003851305_01         33.58%    passed     96.88%\n",
      " 855  Pseudomonas_chlororaphis_003851305_02         33.57%    passed     97.04%\n",
      " 856  Pseudomonas_chlororaphis_003851305_03         33.57%    passed     97.04%\n",
      " 857  Pseudomonas_chlororaphis_003851305_04         33.57%    passed     96.49%\n",
      " 858  Pseudomonas_chlororaphis_001023535_00         33.36%    passed     98.48%\n",
      " 859  Pseudomonas_chlororaphis_001023535_01         30.49%    passed     91.74%\n",
      " 860  Pseudomonas_chlororaphis_001023535_02         33.57%    passed     99.14%\n",
      " 861  Pseudomonas_chlororaphis_001023535_04         33.41%    passed     99.21%\n",
      " 862  Pseudomonas_chlororaphis_001023535_05         32.97%    passed     94.68%\n",
      " 863  Pseudomonas_aeruginosa_000473745_01           33.84%    passed     99.04%\n",
      " 864  Pseudomonas_aeruginosa_000473745_03           33.84%    passed     99.04%\n",
      " 865  Pseudomonas_syringae_001913235_00             32.79%    passed     97.23%\n",
      " 866  Pseudomonas_syringae_001913235_01             32.79%    passed     97.23%\n",
      " 867  Pseudomonas_syringae_001913235_02             32.79%    passed     97.63%\n",
      " 868  Pseudomonas_syringae_001913235_03             32.79%    passed     97.23%\n",
      " 869  Pseudomonas_syringae_001913235_04             29.92%    passed     98.24%\n",
      " 870  Pseudomonas_aeruginosa_001515915_00           33.84%    passed     98.29%\n",
      " 871  Pseudomonas_aeruginosa_001515915_01           33.83%    passed     98.15%\n",
      " 872  Pseudomonas_aeruginosa_001515915_02           33.83%    passed     98.81%\n",
      " 873  Pseudomonas_aeruginosa_001515915_03           33.81%    passed     98.27%\n",
      " 874  Pseudomonas_protegens_002208745_00            33.23%    passed     80.91%\n",
      " 875  Pseudomonas_protegens_002208745_01            30.28%    passed     61.24%\n",
      " 876  Pseudomonas_protegens_002208745_02            33.22%    passed     80.01%\n",
      " 877  Pseudomonas_protegens_002208745_03            33.25%    passed     81.94%\n",
      " 878  Pseudomonas_protegens_002208745_04            33.23%    passed     79.25%\n",
      " 879  Pseudomonas_aeruginosa_000829885_00           33.81%    passed     96.93%\n",
      " 880  Pseudomonas_aeruginosa_000829885_01           33.81%    passed     96.93%\n",
      " 881  Pseudomonas_aeruginosa_000829885_02           33.83%    passed     98.26%\n",
      " 882  Pseudomonas_aeruginosa_000829885_03           33.83%    passed     97.35%\n",
      " 883  Pseudomonas_aeruginosa_003193645_00           33.85%    passed     97.50%\n",
      " 884  Pseudomonas_aeruginosa_003193645_01           33.85%    passed     97.86%\n",
      " 885  Pseudomonas_aeruginosa_003193645_02           33.85%    passed     97.86%\n",
      " 886  Pseudomonas_aeruginosa_003193645_03           33.84%    passed     97.53%\n",
      " 887  Pseudomonas_aeruginosa_002192495_00           33.83%    passed     97.61%\n",
      " 888  Pseudomonas_aeruginosa_002192495_01           33.83%    passed     98.15%\n",
      " 889  Pseudomonas_aeruginosa_002192495_02           33.83%    passed     97.61%\n",
      " 890  Pseudomonas_aeruginosa_002192495_03           33.85%    passed     98.33%\n",
      " 891  Pseudomonas_aeruginosa_000017205_00           33.85%    passed     98.19%\n",
      " 892  Pseudomonas_aeruginosa_000017205_01           33.85%    passed     98.19%\n",
      " 893  Pseudomonas_aeruginosa_000017205_02           33.85%    passed     97.75%\n",
      " 894  Pseudomonas_aeruginosa_000017205_03           33.85%    passed     97.88%\n",
      " 895  Pseudomonas_aeruginosa_900637045_00           33.83%    passed     98.77%\n",
      " 896  Pseudomonas_aeruginosa_900637045_01           33.83%    passed     98.55%\n",
      " 897  Pseudomonas_aeruginosa_900637045_02           33.83%    passed     98.77%\n",
      " 898  Pseudomonas_aeruginosa_900637045_03           33.83%    passed     98.77%\n",
      " 899  Pseudomonas_aeruginosa_000414035_00           33.83%    passed     97.94%\n",
      " 900  Pseudomonas_aeruginosa_000414035_01           33.81%    passed     98.03%\n",
      " 901  Pseudomonas_aeruginosa_000414035_02           33.81%    passed     98.03%\n",
      " 902  Pseudomonas_aeruginosa_000414035_03           34.22%    passed     98.05%\n",
      " 903  Pseudomonas_stutzeri_003952645_00             30.04%    passed     83.85%\n",
      " 904  Pseudomonas_stutzeri_003952645_01             32.91%    passed     92.63%\n",
      " 905  Pseudomonas_stutzeri_003952645_02             32.73%    passed     85.62%\n",
      " 906  Pseudomonas_stutzeri_003952645_03             32.91%    passed     92.63%\n",
      " 907  Pseudomonas_stutzeri_003952645_04             32.75%    passed     83.90%\n",
      " 908  Pseudomonas_stutzeri_003952645_05             35.45%    passed     80.78%\n",
      " 909  Pseudomonas_stutzeri_003952645_06             32.91%    passed     93.36%\n",
      " 910  Pseudomonas_aeruginosa_003716765_00           33.76%    passed     84.72%\n",
      " 911  Pseudomonas_aeruginosa_003716765_01           33.86%    passed     98.68%\n",
      " 912  Pseudomonas_aeruginosa_003716765_02           33.81%    passed     97.33%\n",
      " 913  Pseudomonas_aeruginosa_003716765_03           33.79%    passed     97.44%\n",
      " 914  Pseudomonas_frederiksbergensis_002355315_00   33.12%    passed     99.14%\n",
      " 915  Pseudomonas_frederiksbergensis_002355315_01   32.17%    passed     95.44%\n",
      " 916  Pseudomonas_frederiksbergensis_002355315_02   32.91%    passed     99.22%\n",
      " 917  Pseudomonas_frederiksbergensis_002355315_03   32.18%    passed     95.50%\n",
      " 918  Pseudomonas_frederiksbergensis_002355315_04   32.17%    passed     95.44%\n",
      " 919  Pseudomonas_frederiksbergensis_002355315_05   32.77%    passed     91.80%\n",
      " 920  Pseudomonas_aeruginosa_000226155_00           33.85%    passed     98.47%\n",
      " 921  Pseudomonas_aeruginosa_000226155_01           33.85%    passed     98.47%\n",
      " 922  Pseudomonas_aeruginosa_000226155_02           33.85%    passed     98.47%\n",
      " 923  Pseudomonas_aeruginosa_000226155_03           33.85%    passed     98.47%\n",
      "WARNING: 7 sequences contain more than 50% gaps/ambiguity\n",
      "****  TOTAL                                         33.59%  0 sequences failed composition chi2 test (p-value<5%; df=3)\n",
      "NOTE: 257 identical sequences will be ignored during tree search\n",
      "\n",
      "Create initial parsimony tree by phylogenetic likelihood library (PLL)... 0.319 seconds\n",
      "\n",
      "NOTE: 144 MB RAM is required!\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "1. Initial log-likelihood: -59441.073\n",
      "2. Current log-likelihood: -54268.379\n",
      "3. Current log-likelihood: -53813.480\n",
      "4. Current log-likelihood: -53811.548\n",
      "Optimal log-likelihood: -53811.487\n",
      "Rate parameters:  A-C: 1.000  A-G: 2.881  A-T: 1.000  C-G: 1.000  C-T: 2.881  G-T: 1.000\n",
      "Base frequencies:   A: 0.261  C: 0.219  G: 0.302  T: 0.218\n",
      "Gamma shape alpha: 0.161\n",
      "Parameters optimization took 4 rounds (2.482 sec)\n",
      "\n",
      "--------------------------------------------------------------------\n",
      "|             INITIALIZING CANDIDATE TREE SET                      |\n",
      "--------------------------------------------------------------------\n",
      "Computing ML distances based on estimated model parameters... 8.076 sec\n",
      "Computing BIONJ tree...\n",
      "0.446 seconds\n",
      "Log-likelihood of BIONJ tree: -55064.511\n",
      "Generating 98 parsimony trees... 34.822 seconds 100 distinct starting trees\n",
      "Computing log-likelihood of 100 initial trees ... 39.571 seconds\n",
      "Optimizing top 20 initial trees with NNI...\n",
      "Iteration 1 / LogL: -53643.863 / Time: 0h:1m:21s\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "BETTER TREE FOUND at iteration 1: -53643.794\n",
      "Iteration 2 / LogL: -53621.187 / Time: 0h:1m:23s\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "BETTER TREE FOUND at iteration 2: -53620.973\n",
      "Iteration 3 / LogL: -53631.446 / Time: 0h:1m:27s\n",
      "Iteration 4 / LogL: -53669.279 / Time: 0h:1m:33s\n",
      "Iteration 5 / LogL: -53669.476 / Time: 0h:1m:35s\n",
      "Iteration 6 / LogL: -53713.280 / Time: 0h:1m:37s\n",
      "Iteration 7 / LogL: -53688.984 / Time: 0h:1m:39s\n",
      "Iteration 8 / LogL: -53610.775 / Time: 0h:1m:42s\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "BETTER TREE FOUND at iteration 8: -53610.715\n",
      "Iteration 9 / LogL: -53683.731 / Time: 0h:1m:46s\n",
      "Iteration 10 / LogL: -53672.747 / Time: 0h:1m:47s\n",
      "Iteration 11 / LogL: -53681.166 / Time: 0h:1m:49s\n",
      "Iteration 12 / LogL: -53661.419 / Time: 0h:1m:50s\n",
      "Iteration 13 / LogL: -53658.682 / Time: 0h:1m:52s\n",
      "Iteration 14 / LogL: -53686.689 / Time: 0h:1m:54s\n",
      "Iteration 15 / LogL: -53701.104 / Time: 0h:1m:55s\n",
      "Iteration 16 / LogL: -53709.990 / Time: 0h:1m:57s\n",
      "Iteration 17 / LogL: -53687.064 / Time: 0h:1m:59s\n",
      "Iteration 18 / LogL: -53696.392 / Time: 0h:2m:0s\n",
      "Iteration 19 / LogL: -53733.763 / Time: 0h:2m:2s\n",
      "Iteration 20 / LogL: -53671.125 / Time: 0h:2m:4s\n",
      "Average CPU time for 1 NNI search: 16.748\n",
      "Finish initializing candidate tree set. Number of distinct locally optimal trees: 20\n",
      "Current best tree score: -53610.715 / CPU time: 662.672\n",
      "--------------------------------------------------------------------\n",
      "|               OPTIMIZING CANDIDATE TREE SET                      |\n",
      "--------------------------------------------------------------------\n",
      "Iteration 30 / LogL: -53665.578 / Time: 0h:2m:30s (0h:6m:8s left)\n",
      "Iteration 40 / LogL: -53804.485 / Time: 0h:2m:55s (0h:4m:34s left)\n",
      "Iteration 50 / LogL: -53852.261 / Time: 0h:3m:21s (0h:3m:29s left)\n",
      "Iteration 60 / LogL: -53639.157 / Time: 0h:3m:52s (0h:2m:41s left)\n",
      "Iteration 70 / LogL: -53805.293 / Time: 0h:4m:16s (0h:1m:55s left)\n",
      "Iteration 80 / LogL: -53665.327 / Time: 0h:4m:58s (0h:1m:19s left)\n",
      "Estimate model parameters (epsilon = 0.010)\n",
      "BETTER TREE FOUND at iteration 86: -53593.369\n",
      "Iteration 90 / LogL: -53953.464 / Time: 0h:5m:27s (0h:5m:53s left)\n",
      "Iteration 100 / LogL: -53644.263 / Time: 0h:6m:13s (0h:5m:23s left)\n",
      "Iteration 110 / LogL: -53761.446 / Time: 0h:6m:45s (0h:4m:42s left)\n",
      "Iteration 120 / LogL: -53789.794 / Time: 0h:7m:15s (0h:4m:1s left)\n",
      "Iteration 130 / LogL: -53714.681 / Time: 0h:7m:57s (0h:3m:27s left)\n",
      "Iteration 140 / LogL: -53640.258 / Time: 0h:8m:36s (0h:2m:50s left)\n",
      "Iteration 150 / LogL: -53630.313 / Time: 0h:9m:2s (0h:2m:10s left)\n",
      "Iteration 160 / LogL: -53982.217 / Time: 0h:9m:26s (0h:1m:32s left)\n",
      "Iteration 170 / LogL: -53782.119 / Time: 0h:9m:50s (0h:0m:55s left)\n",
      "Iteration 180 / LogL: -53734.150 / Time: 0h:10m:14s (0h:0m:20s left)\n",
      "TREE SEARCH COMPLETED AFTER 186 ITERATIONS / Time: 0h:10m:29s\n",
      "\n",
      "--------------------------------------------------------------------\n",
      "|                    FINALIZING TREE SEARCH                        |\n",
      "--------------------------------------------------------------------\n",
      "Performs final model parameters optimization\n",
      "Estimate model parameters (epsilon = 0.010)\n",
      "1. Initial log-likelihood: -53593.356\n",
      "2. Current log-likelihood: -53593.333\n",
      "Optimal log-likelihood: -53593.333\n",
      "Rate parameters:  A-C: 1.000  A-G: 2.892  A-T: 1.000  C-G: 1.000  C-T: 2.892  G-T: 1.000\n",
      "Base frequencies:   A: 0.261  C: 0.219  G: 0.302  T: 0.218\n",
      "Gamma shape alpha: 0.161\n",
      "Parameters optimization took 2 rounds (0.802 sec)\n",
      "\n",
      "BEST SCORE FOUND : -53593.333\n",
      "Total tree length: 4.836\n",
      "\n",
      "Total number of iterations: 186\n",
      "CPU time used for tree search: 4480.776 sec (1h:14m:40s)\n",
      "Wall-clock time used for tree search: 623.517 sec (0h:10m:23s)\n",
      "Total CPU time used: 4524.552 sec (1h:15m:24s)\n",
      "Total wall-clock time used: 630.066 sec (0h:10m:30s)\n",
      "\n",
      "Analysis results written to: \n",
      "  IQ-TREE report:                ./014_results/Pseudomonas.fasta.iqtree\n",
      "  Maximum-likelihood tree:       ./014_results/Pseudomonas.fasta.treefile\n",
      "  Likelihood distances:          ./014_results/Pseudomonas.fasta.mldist\n",
      "  Screen log file:               ./014_results/Pseudomonas.fasta.log\n",
      "\n",
      "Date and Time: Mon Feb 25 09:56:20 2019\n",
      "\n"
     ]
    }
   ],
   "source": [
    "outcheck = subprocess.check_output(\"muscle -maketree -in \" + outdir + \".fasta -out \" + outdir + \"_upgma.treefile\", shell=True,universal_newlines=True)\n",
    "outcheck = subprocess.check_output(\"iqtree-omp -s \" + outdir + \".fasta -nt 8 -m HKY+G\",shell=True,universal_newlines=True)\n",
    "print (outcheck)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def patristic_distances_from_treefile (filename, have_paralogs = False, shuffle = False):\n",
    "    tree = dendropy.Tree.get(path=filename, schema=\"newick\", preserve_underscores=True)\n",
    "    if shuffle:\n",
    "        tree = dendropy.simulate.treesim.pure_kingman_tree(taxon_namespace=tree.taxon_namespace, pop_size = 1000)\n",
    "    species = ['_'.join(t.label.split('_')[:2]) for t in tree.taxon_namespace] ## follow order of taxon_namespace\n",
    "    genus = [t.label.split('_')[0] for t in tree.taxon_namespace]\n",
    "    ntaxa = len(tree.taxon_namespace)\n",
    "    distmat = np.zeros((ntaxa,ntaxa)) # diagonals are zero\n",
    "    nodemat = np.zeros((ntaxa,ntaxa))\n",
    "    pdm = tree.phylogenetic_distance_matrix() # initialises class\n",
    "    for i,j in itertools.combinations(range(ntaxa),2):\n",
    "        distmat[i,j] = distmat[j,i] = pdm.distance(tree.taxon_namespace[i], tree.taxon_namespace[j])\n",
    "        nodemat[i,j] = nodemat[j,i] = pdm.path_edge_count(tree.taxon_namespace[i], tree.taxon_namespace[j])\n",
    "    if have_paralogs:\n",
    "        sample = ['_'.join(t.label.split('_')[:3]) for t in tree.taxon_namespace]\n",
    "        return species, genus, distmat, nodemat, sample\n",
    "    return species, genus, distmat, nodemat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def silhouette_str(dist, labels):\n",
    "    distrib = metrics.silhouette_samples(dist, labels, metric=\"precomputed\")\n",
    "    return '{:7.3f}'.format(np.percentile(distrib, 5)) + \\\n",
    "    '{:7.3f}'.format(np.percentile(distrib, 25)) + \\\n",
    "    '{:7.3f}  '.format(sum(distrib>0)/float(len(distrib))) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Each triplet from `silhouette_str()` has $5\\%$, $25\\%$ and proportion of positive silhouettes. And we have two triplets on the left for _species_ (S), and two on the right for _strain_ (T), one being from weighted (W) and one from unweighted (U) patristic distances. Therefore the header is: SW SU TW TU\n",
    "\n",
    "By the way having low silhouettte scores is not necessarily bad, since\n",
    "1. clusters of size one will have score of _zero_ **by definition** (should be _one_ given the silhouette function but this would lead to ever-increasing clusterings)\n",
    "2. clusters with within-distance zero are perfect, while for phylogenies we want non-zero diversification\n",
    "3. they can be negative if for a case like `((A:a,A:a,A:a),B:b)` with $a>b$ since $d(A,A)=2a > a+b = d(A,B)$ --- however the taxonomic classification is fine for A and B"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "random \t  -0.536 -0.427  0.005    -0.192 -0.142  0.012    |   -0.963 -0.792  0.005    -0.554 -0.444  0.012  \n",
      "upgma  \t  -0.475  0.527  0.852    -0.493  0.054  0.782    |   -0.978 -0.282  0.482    -0.729 -0.319  0.535  \n",
      "ML     \t  -0.675  0.573  0.815    -0.524 -0.078  0.716    |   -0.932 -0.465  0.450    -0.697 -0.321  0.492  \n"
     ]
    }
   ],
   "source": [
    "sp, ge, mat, mat2, strain = patristic_distances_from_treefile (outdir + \".fasta.treefile\", have_paralogs = True, shuffle = True)\n",
    "print (\"random \\t\", silhouette_str(mat, sp),  silhouette_str(mat2, sp), \" | \", \n",
    "      silhouette_str(mat, strain),  silhouette_str(mat2, strain))\n",
    "\n",
    "for suffix1, title1 in zip([\"_upgma.treefile\",\".fasta.treefile\"], [\"upgma \", \"ML    \"]):\n",
    "    ## operon doenst have _long or _consensus\n",
    "    sp, ge, mat, mat2, strain = patristic_distances_from_treefile (outdir + suffix1, have_paralogs = True)\n",
    "    print (title1, \"\\t\", silhouette_str(mat, sp),  silhouette_str(mat2, sp), \" | \", \n",
    "      silhouette_str(mat, strain),  silhouette_str(mat2, strain))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sp_count_from_list (labels):\n",
    "    x = ['_'.join(t.split('_')[:2]) for t in labels]\n",
    "    return len(set(x))\n",
    "\n",
    "def strain_count_from_list (labels):\n",
    "    x = ['_'.join(t.split('_')[:3]) for t in labels]\n",
    "    return len(set(x))\n",
    "\n",
    "def bipartitions_from_treefile (filename, shuffle = False):\n",
    "    tree = dendropy.Tree.get(path=filename, schema=\"newick\", preserve_underscores=True)\n",
    "    if shuffle:\n",
    "        tree = dendropy.simulate.treesim.pure_kingman_tree(taxon_namespace=tree.taxon_namespace, pop_size = 1000)\n",
    "    tree.encode_bipartitions()\n",
    "    # bits[i] have the list of labels associated to bipartition (edge) i\n",
    "    bits = [[x.label for x in tree.taxon_namespace.bitmask_taxa_list(i.leafset_bitmask)] for i in tree.bipartition_encoding]\n",
    "    sp_count = [sp_count_from_list(i) for i in bits if len(i) > 1]\n",
    "    strain_count = [strain_count_from_list(i) for i in bits if len(i) > 1]\n",
    "    return sp_count, strain_count\n",
    "\n",
    "def bipart_str(counts):\n",
    "    distrib = np.array(counts)\n",
    "    return '{:7.3f}'.format(np.mean(distrib)) + \\\n",
    "    '{:7.3f}'.format(np.percentile(distrib, 90)) + \\\n",
    "    '{:7.3f}\\t'.format(sum(distrib == 1)/float(len(distrib))) \n",
    "\n",
    "sp0, ge0 = bipartitions_from_treefile (outdir + \".fasta.treefile\", shuffle = True)\n",
    "print ('{:7s}'.format(\"random\"), \"\\t\", bipart_str(sp0), bipart_str(ge0))\n",
    "\n",
    "for suffix1, title1 in zip([\"_upgma.treefile\",\".fasta.treefile\"], [\"upgma \", \"ML    \"]):\n",
    "    sp, ge = bipartitions_from_treefile (outdir + suffix1)    \n",
    "    print ('{:7s}'.format(title1), \"\\t\", bipart_str(sp), bipart_str(ge))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
